// @qwik-ui/headless/components/dropdown/dropdown-item.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const useDropdownItem = require("./use-dropdown-item.qwik.cjs");
const HDropdownItem = qwik.component$((props) => {
  const { _index, disabled, closeOnSelect = true, ...rest } = props;
  const { handleClick$, handleKeyDown$, handlePointerOver$, itemId, itemRef, isHighlightedSig } = useDropdownItem.useDropdownItem({
    closeOnSelect,
    ...props
  });
  const handleKeyDownSync$ = qwik.sync$((e) => {
    const keys = [
      "ArrowUp",
      "ArrowDown",
      "ArrowLeft",
      "Home",
      "End",
      "Enter",
      " "
    ];
    if (keys.includes(e.key)) {
      e.preventDefault();
    }
  });
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    ...rest,
    role: "menuitem",
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    onKeyDown$: [
      handleKeyDownSync$,
      handleKeyDown$,
      props.onKeyDown$
    ],
    onPointerOver$: [
      handlePointerOver$,
      props.onPointerOver$
    ],
    tabIndex: -1,
    id: itemId,
    ref: itemRef,
    "aria-disabled": disabled === true ? "true" : "false",
    "data-disabled": disabled,
    "data-highlighted": isHighlightedSig.value,
    "data-menu-item": true,
    "data-close-on-select": closeOnSelect,
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HDropdownItem = HDropdownItem;
