// @qwik-ui/headless/components/dropdown/dropdown-root.qwik.mjs
import { jsx } from "@builder.io/qwik/jsx-runtime";
import { component$, useSignal, useId, useComputed$, useContextProvider, useTask$, Slot } from "@builder.io/qwik";
import { dropdownContextId } from "./dropdown-context.qwik.mjs";
const HDropdownImpl = component$((props) => {
  const { _itemsMap, onOpenChange$, scrollOptions: givenScrollOptions, loop: givenLoop, ...rest } = props;
  const rootRef = useSignal();
  const triggerRef = useSignal();
  const panelRef = useSignal();
  const loop = givenLoop ?? false;
  const localId = useId();
  const dropdownId = `${localId}-dropdown`;
  const itemsMapSig = useComputed$(() => {
    return props._itemsMap;
  });
  const highlightedIndexSig = useSignal(null);
  const isOpenSig = useSignal(false);
  const scrollOptions = givenScrollOptions ?? {
    behavior: "instant",
    block: "nearest"
  };
  const currDisplayValueSig = useSignal();
  const initialLoadSig = useSignal(true);
  const highlightedItemRef = useSignal();
  const isMouseOverPopupSig = useSignal(false);
  const context = {
    itemsMapSig,
    currDisplayValueSig,
    highlightedIndexSig,
    isOpenSig,
    triggerRef,
    panelRef,
    highlightedItemRef,
    isMouseOverPopupSig,
    localId,
    scrollOptions,
    loop
  };
  useContextProvider(dropdownContextId, context);
  useTask$(function reactiveUserOpen({ track }) {
    const bindOpenSig = props["bind:open"];
    if (!bindOpenSig) return;
    track(() => bindOpenSig.value);
    isOpenSig.value = bindOpenSig.value ?? isOpenSig.value;
  });
  useTask$(function onOpenChangeTask({ track }) {
    track(() => isOpenSig.value);
    if (!initialLoadSig.value) {
      onOpenChange$?.(isOpenSig.value);
    }
  });
  const activeDescendantSig = useComputed$(() => {
    if (!isOpenSig.value) {
      return "";
    }
    const highlightedIndex = context.highlightedIndexSig.value ?? -1;
    const highlightedItem = context.itemsMapSig.value.get(highlightedIndex);
    if (highlightedIndex === null || highlightedIndex === -1 || highlightedItem?.disabled) {
      return "";
    }
    return `${context.localId}-${highlightedIndex}`;
  });
  useTask$(() => {
    initialLoadSig.value = false;
  });
  return /* @__PURE__ */ jsx("div", {
    role: "group",
    "data-qui-dropdown": true,
    ref: rootRef,
    "data-open": context.isOpenSig.value ? true : void 0,
    "data-closed": !context.isOpenSig.value ? true : void 0,
    "aria-controls": dropdownId,
    "aria-expanded": context.isOpenSig.value,
    "aria-haspopup": "true",
    "aria-activedescendant": activeDescendantSig.value,
    ...rest,
    children: /* @__PURE__ */ jsx(Slot, {})
  });
});
export {
  HDropdownImpl
};
