// @qwik-ui/headless/components/dropdown/dropdown-checkbox-item.qwik.mjs
import { jsx } from "@builder.io/qwik/jsx-runtime";
import { component$, useSignal, useTask$, $, sync$, Slot } from "@builder.io/qwik";
import { CheckboxRoot } from "../checkbox/checkbox-root.qwik.mjs";
import { useDropdownItem } from "./use-dropdown-item.qwik.mjs";
const HDropdownCheckboxItem = component$((props) => {
  const { disabled, onChange$, closeOnSelect = false, ...rest } = props;
  const checkedSig = useSignal(false);
  const checkboxRef = useSignal();
  useTask$(function reactiveUserChecked({ track }) {
    const bindCheckedSig = props["bind:checked"];
    if (!bindCheckedSig) return;
    track(() => bindCheckedSig.value);
    checkedSig.value = bindCheckedSig.value ?? checkedSig.value;
  });
  useTask$(function onChangeTask({ track }) {
    track(() => checkedSig.value);
    onChange$?.(checkedSig.value);
  });
  const toggleChecked$ = $(() => {
    checkedSig.value = !checkedSig.value;
  });
  const { handleClick$, handleKeyDown$, handlePointerOver$, itemId, itemRef, isHighlightedSig } = useDropdownItem({
    ...props,
    onItemSelect: toggleChecked$,
    closeOnSelect
  });
  const handleKeyDownSync$ = sync$((e) => {
    const keys = [
      "ArrowUp",
      "ArrowDown",
      "ArrowLeft",
      "Home",
      "End",
      "Enter",
      " "
    ];
    if (keys.includes(e.key)) {
      e.preventDefault();
    }
  });
  return /* @__PURE__ */ jsx("div", {
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    tabIndex: -1,
    id: itemId,
    onKeyDown$: [
      handleKeyDownSync$,
      handleKeyDown$,
      props.onKeyDown$
    ],
    onPointerOver$: [
      handlePointerOver$,
      props.onPointerOver$
    ],
    ref: itemRef,
    "aria-disabled": disabled === true ? "true" : "false",
    "data-disabled": disabled,
    "aria-checked": checkedSig.value ? "true" : "false",
    "data-highlighted": isHighlightedSig.value,
    "data-checked": checkedSig.value,
    "data-close-on-select": props.closeOnSelect,
    "data-menu-item": true,
    children: /* @__PURE__ */ jsx(CheckboxRoot, {
      "bind:checked": checkedSig,
      "preventdefault:click": true,
      ref: checkboxRef,
      style: {
        pointerEvents: "none"
      },
      ...rest,
      children: /* @__PURE__ */ jsx(Slot, {})
    })
  });
});
export {
  HDropdownCheckboxItem
};
