// @qwik-ui/headless/components/dropdown/dropdown-root.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const dropdownContext = require("./dropdown-context.qwik.cjs");
const HDropdownImpl = qwik.component$((props) => {
  const { _itemsMap, onOpenChange$, scrollOptions: givenScrollOptions, loop: givenLoop, ...rest } = props;
  const rootRef = qwik.useSignal();
  const triggerRef = qwik.useSignal();
  const panelRef = qwik.useSignal();
  const loop = givenLoop ?? false;
  const localId = qwik.useId();
  const dropdownId = `${localId}-dropdown`;
  const itemsMapSig = qwik.useComputed$(() => {
    return props._itemsMap;
  });
  const highlightedIndexSig = qwik.useSignal(null);
  const isOpenSig = qwik.useSignal(false);
  const scrollOptions = givenScrollOptions ?? {
    behavior: "instant",
    block: "nearest"
  };
  const currDisplayValueSig = qwik.useSignal();
  const initialLoadSig = qwik.useSignal(true);
  const highlightedItemRef = qwik.useSignal();
  const isMouseOverPopupSig = qwik.useSignal(false);
  const context = {
    itemsMapSig,
    currDisplayValueSig,
    highlightedIndexSig,
    isOpenSig,
    triggerRef,
    panelRef,
    highlightedItemRef,
    isMouseOverPopupSig,
    localId,
    scrollOptions,
    loop
  };
  qwik.useContextProvider(dropdownContext.dropdownContextId, context);
  qwik.useTask$(function reactiveUserOpen({ track }) {
    const bindOpenSig = props["bind:open"];
    if (!bindOpenSig) return;
    track(() => bindOpenSig.value);
    isOpenSig.value = bindOpenSig.value ?? isOpenSig.value;
  });
  qwik.useTask$(function onOpenChangeTask({ track }) {
    track(() => isOpenSig.value);
    if (!initialLoadSig.value) {
      onOpenChange$?.(isOpenSig.value);
    }
  });
  const activeDescendantSig = qwik.useComputed$(() => {
    if (!isOpenSig.value) {
      return "";
    }
    const highlightedIndex = context.highlightedIndexSig.value ?? -1;
    const highlightedItem = context.itemsMapSig.value.get(highlightedIndex);
    if (highlightedIndex === null || highlightedIndex === -1 || highlightedItem?.disabled) {
      return "";
    }
    return `${context.localId}-${highlightedIndex}`;
  });
  qwik.useTask$(() => {
    initialLoadSig.value = false;
  });
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    role: "group",
    "data-qui-dropdown": true,
    ref: rootRef,
    "data-open": context.isOpenSig.value ? true : void 0,
    "data-closed": !context.isOpenSig.value ? true : void 0,
    "aria-controls": dropdownId,
    "aria-expanded": context.isOpenSig.value,
    "aria-haspopup": "true",
    "aria-activedescendant": activeDescendantSig.value,
    ...rest,
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HDropdownImpl = HDropdownImpl;
