// @qwik-ui/headless/components/dropdown/dropdown-item.qwik.mjs
import { jsx } from "@builder.io/qwik/jsx-runtime";
import { component$, sync$, Slot } from "@builder.io/qwik";
import { useDropdownItem } from "./use-dropdown-item.qwik.mjs";
const HDropdownItem = component$((props) => {
  const { _index, disabled, closeOnSelect = true, ...rest } = props;
  const { handleClick$, handleKeyDown$, handlePointerOver$, itemId, itemRef, isHighlightedSig } = useDropdownItem({
    closeOnSelect,
    ...props
  });
  const handleKeyDownSync$ = sync$((e) => {
    const keys = [
      "ArrowUp",
      "ArrowDown",
      "ArrowLeft",
      "Home",
      "End",
      "Enter",
      " "
    ];
    if (keys.includes(e.key)) {
      e.preventDefault();
    }
  });
  return /* @__PURE__ */ jsx("div", {
    ...rest,
    role: "menuitem",
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    onKeyDown$: [
      handleKeyDownSync$,
      handleKeyDown$,
      props.onKeyDown$
    ],
    onPointerOver$: [
      handlePointerOver$,
      props.onPointerOver$
    ],
    tabIndex: -1,
    id: itemId,
    ref: itemRef,
    "aria-disabled": disabled === true ? "true" : "false",
    "data-disabled": disabled,
    "data-highlighted": isHighlightedSig.value,
    "data-menu-item": true,
    "data-close-on-select": closeOnSelect,
    children: /* @__PURE__ */ jsx(Slot, {})
  });
});
export {
  HDropdownItem
};
