// @qwik-ui/headless/components/dropdown/dropdown-checkbox-item.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const checkboxRoot = require("../checkbox/checkbox-root.qwik.cjs");
const useDropdownItem = require("./use-dropdown-item.qwik.cjs");
const HDropdownCheckboxItem = qwik.component$((props) => {
  const { disabled, onChange$, closeOnSelect = false, ...rest } = props;
  const checkedSig = qwik.useSignal(false);
  const checkboxRef = qwik.useSignal();
  qwik.useTask$(function reactiveUserChecked({ track }) {
    const bindCheckedSig = props["bind:checked"];
    if (!bindCheckedSig) return;
    track(() => bindCheckedSig.value);
    checkedSig.value = bindCheckedSig.value ?? checkedSig.value;
  });
  qwik.useTask$(function onChangeTask({ track }) {
    track(() => checkedSig.value);
    onChange$?.(checkedSig.value);
  });
  const toggleChecked$ = qwik.$(() => {
    checkedSig.value = !checkedSig.value;
  });
  const { handleClick$, handleKeyDown$, handlePointerOver$, itemId, itemRef, isHighlightedSig } = useDropdownItem.useDropdownItem({
    ...props,
    onItemSelect: toggleChecked$,
    closeOnSelect
  });
  const handleKeyDownSync$ = qwik.sync$((e) => {
    const keys = [
      "ArrowUp",
      "ArrowDown",
      "ArrowLeft",
      "Home",
      "End",
      "Enter",
      " "
    ];
    if (keys.includes(e.key)) {
      e.preventDefault();
    }
  });
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    tabIndex: -1,
    id: itemId,
    onKeyDown$: [
      handleKeyDownSync$,
      handleKeyDown$,
      props.onKeyDown$
    ],
    onPointerOver$: [
      handlePointerOver$,
      props.onPointerOver$
    ],
    ref: itemRef,
    "aria-disabled": disabled === true ? "true" : "false",
    "data-disabled": disabled,
    "aria-checked": checkedSig.value ? "true" : "false",
    "data-highlighted": isHighlightedSig.value,
    "data-checked": checkedSig.value,
    "data-close-on-select": props.closeOnSelect,
    "data-menu-item": true,
    children: /* @__PURE__ */ jsxRuntime.jsx(checkboxRoot.CheckboxRoot, {
      "bind:checked": checkedSig,
      "preventdefault:click": true,
      ref: checkboxRef,
      style: {
        pointerEvents: "none"
      },
      ...rest,
      children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
    })
  });
});
exports.HDropdownCheckboxItem = HDropdownCheckboxItem;
