// @qwik-ui/headless/components/dropdown/dropdown-trigger.qwik.mjs
import { jsx } from "@builder.io/qwik/jsx-runtime";
import { component$, useContext, useSignal, $, sync$, Slot } from "@builder.io/qwik";
import { dropdownContextId } from "./dropdown-context.qwik.mjs";
import { useDropdown } from "./use-dropdown.qwik.mjs";
const HDropdownTrigger = component$((props) => {
  const context = useContext(dropdownContextId);
  const { getNextEnabledItemIndex$ } = useDropdown();
  const triggerId = `${context.localId}-trigger`;
  const isInitialKeyDownSig = useSignal(true);
  const handleClick$ = $(() => {
    context.isOpenSig.value = !context.isOpenSig.value;
  });
  const handleKeyDownSync$ = sync$((e) => {
    const keys = [
      "ArrowUp",
      "ArrowDown",
      "ArrowRight",
      "ArrowLeft",
      "Home",
      "End",
      "PageDown",
      "PageUp",
      "Enter",
      " "
    ];
    if (keys.includes(e.key)) {
      e.preventDefault();
    }
  });
  const handleKeyDown$ = $(async (e) => {
    if (!context.itemsMapSig.value) return;
    switch (e.key) {
      case "Tab":
      case "Escape":
        context.isOpenSig.value = false;
        break;
      case "ArrowDown":
      case "ArrowUp":
      case "Enter":
      case " ":
        if (!context.isOpenSig.value) {
          context.isOpenSig.value = true;
        }
        break;
    }
    if (context.highlightedIndexSig.value === null) {
      context.highlightedIndexSig.value = await getNextEnabledItemIndex$(-1);
    }
    while (context.highlightedItemRef.value !== document.activeElement) {
      await new Promise((resolve) => setTimeout(resolve, 5));
      context.highlightedItemRef.value?.focus();
    }
    if (!isInitialKeyDownSig.value) return;
  });
  return /* @__PURE__ */ jsx("button", {
    "data-trigger": true,
    ...props,
    id: triggerId,
    ref: context.triggerRef,
    "preventdefault:click": true,
    "preventdefault:blur": true,
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    onKeyDown$: [
      handleKeyDownSync$,
      handleKeyDown$,
      props.onKeyDown$
    ],
    "data-open": context.isOpenSig.value ? true : void 0,
    "data-closed": !context.isOpenSig.value ? true : void 0,
    "aria-expanded": context.isOpenSig.value,
    "aria-haspopup": "true",
    children: /* @__PURE__ */ jsx(Slot, {})
  });
});
export {
  HDropdownTrigger
};
