// @qwik-ui/headless/components/dropdown/dropdown-radio-item.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const checkboxRoot = require("../checkbox/checkbox-root.qwik.cjs");
const dropdownContext = require("./dropdown-context.qwik.cjs");
const useDropdownItem = require("./use-dropdown-item.qwik.cjs");
const HDropdownRadioItem = qwik.component$((props) => {
  const { disabled = false, value, closeOnSelect = false, onChange$, ...rest } = props;
  const groupContext = qwik.useContext(dropdownContext.dropdownRadioGroupContextId);
  const checkedSig = qwik.useSignal(value === groupContext.valueSig.value);
  const checkboxRef = qwik.useSignal();
  qwik.useTask$(function trackGroupValue({ track }) {
    track(() => groupContext.valueSig.value);
    checkedSig.value = value === groupContext.valueSig.value;
    onChange$?.(checkedSig.value);
  });
  const handleKeyDownSync$ = qwik.sync$((e) => {
    const keys = [
      "ArrowUp",
      "ArrowDown",
      "ArrowLeft",
      "Home",
      "End",
      "Enter",
      " "
    ];
    if (keys.includes(e.key)) {
      e.preventDefault();
    }
  });
  const onItemSelect$ = qwik.$(() => {
    groupContext.valueSig.value = value;
  });
  const { handleClick$, handleKeyDown$, handlePointerOver$, itemId, itemRef, isHighlightedSig } = useDropdownItem.useDropdownItem({
    ...props,
    onItemSelect: onItemSelect$,
    closeOnSelect
  });
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    tabIndex: -1,
    id: itemId,
    onKeyDown$: [
      handleKeyDownSync$,
      handleKeyDown$,
      props.onKeyDown$
    ],
    onPointerOver$: [
      handlePointerOver$,
      props.onPointerOver$
    ],
    ref: itemRef,
    role: "menuitemradio",
    "aria-checked": checkedSig.value ? "true" : "false",
    "aria-disabled": disabled || groupContext.disabled,
    "data-disabled": disabled || groupContext.disabled,
    "data-highlighted": isHighlightedSig.value,
    "data-checked": checkedSig.value,
    "data-close-on-select": props.closeOnSelect,
    "data-menu-item": true,
    children: /* @__PURE__ */ jsxRuntime.jsx(checkboxRoot.CheckboxRoot, {
      "bind:checked": checkedSig,
      "preventdefault:click": true,
      ref: checkboxRef,
      style: {
        pointerEvents: "none"
      },
      ...rest,
      children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
    })
  });
});
exports.HDropdownRadioItem = HDropdownRadioItem;
