// @qwik-ui/headless/components/dropdown/dropdown-radio-item.qwik.mjs
import { jsx } from "@builder.io/qwik/jsx-runtime";
import { component$, useContext, useSignal, useTask$, sync$, $, Slot } from "@builder.io/qwik";
import { CheckboxRoot } from "../checkbox/checkbox-root.qwik.mjs";
import { dropdownRadioGroupContextId } from "./dropdown-context.qwik.mjs";
import { useDropdownItem } from "./use-dropdown-item.qwik.mjs";
const HDropdownRadioItem = component$((props) => {
  const { disabled = false, value, closeOnSelect = false, onChange$, ...rest } = props;
  const groupContext = useContext(dropdownRadioGroupContextId);
  const checkedSig = useSignal(value === groupContext.valueSig.value);
  const checkboxRef = useSignal();
  useTask$(function trackGroupValue({ track }) {
    track(() => groupContext.valueSig.value);
    checkedSig.value = value === groupContext.valueSig.value;
    onChange$?.(checkedSig.value);
  });
  const handleKeyDownSync$ = sync$((e) => {
    const keys = [
      "ArrowUp",
      "ArrowDown",
      "ArrowLeft",
      "Home",
      "End",
      "Enter",
      " "
    ];
    if (keys.includes(e.key)) {
      e.preventDefault();
    }
  });
  const onItemSelect$ = $(() => {
    groupContext.valueSig.value = value;
  });
  const { handleClick$, handleKeyDown$, handlePointerOver$, itemId, itemRef, isHighlightedSig } = useDropdownItem({
    ...props,
    onItemSelect: onItemSelect$,
    closeOnSelect
  });
  return /* @__PURE__ */ jsx("div", {
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    tabIndex: -1,
    id: itemId,
    onKeyDown$: [
      handleKeyDownSync$,
      handleKeyDown$,
      props.onKeyDown$
    ],
    onPointerOver$: [
      handlePointerOver$,
      props.onPointerOver$
    ],
    ref: itemRef,
    role: "menuitemradio",
    "aria-checked": checkedSig.value ? "true" : "false",
    "aria-disabled": disabled || groupContext.disabled,
    "data-disabled": disabled || groupContext.disabled,
    "data-highlighted": isHighlightedSig.value,
    "data-checked": checkedSig.value,
    "data-close-on-select": props.closeOnSelect,
    "data-menu-item": true,
    children: /* @__PURE__ */ jsx(CheckboxRoot, {
      "bind:checked": checkedSig,
      "preventdefault:click": true,
      ref: checkboxRef,
      style: {
        pointerEvents: "none"
      },
      ...rest,
      children: /* @__PURE__ */ jsx(Slot, {})
    })
  });
});
export {
  HDropdownRadioItem
};
