// @qwik-ui/headless/components/dropdown/dropdown-trigger.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const dropdownContext = require("./dropdown-context.qwik.cjs");
const useDropdown = require("./use-dropdown.qwik.cjs");
const HDropdownTrigger = qwik.component$((props) => {
  const context = qwik.useContext(dropdownContext.dropdownContextId);
  const { getNextEnabledItemIndex$ } = useDropdown.useDropdown();
  const triggerId = `${context.localId}-trigger`;
  const isInitialKeyDownSig = qwik.useSignal(true);
  const handleClick$ = qwik.$(() => {
    context.isOpenSig.value = !context.isOpenSig.value;
  });
  const handleKeyDownSync$ = qwik.sync$((e) => {
    const keys = [
      "ArrowUp",
      "ArrowDown",
      "ArrowRight",
      "ArrowLeft",
      "Home",
      "End",
      "PageDown",
      "PageUp",
      "Enter",
      " "
    ];
    if (keys.includes(e.key)) {
      e.preventDefault();
    }
  });
  const handleKeyDown$ = qwik.$(async (e) => {
    if (!context.itemsMapSig.value) return;
    switch (e.key) {
      case "Tab":
      case "Escape":
        context.isOpenSig.value = false;
        break;
      case "ArrowDown":
      case "ArrowUp":
      case "Enter":
      case " ":
        if (!context.isOpenSig.value) {
          context.isOpenSig.value = true;
        }
        break;
    }
    if (context.highlightedIndexSig.value === null) {
      context.highlightedIndexSig.value = await getNextEnabledItemIndex$(-1);
    }
    while (context.highlightedItemRef.value !== document.activeElement) {
      await new Promise((resolve) => setTimeout(resolve, 5));
      context.highlightedItemRef.value?.focus();
    }
    if (!isInitialKeyDownSig.value) return;
  });
  return /* @__PURE__ */ jsxRuntime.jsx("button", {
    "data-trigger": true,
    ...props,
    id: triggerId,
    ref: context.triggerRef,
    "preventdefault:click": true,
    "preventdefault:blur": true,
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    onKeyDown$: [
      handleKeyDownSync$,
      handleKeyDown$,
      props.onKeyDown$
    ],
    "data-open": context.isOpenSig.value ? true : void 0,
    "data-closed": !context.isOpenSig.value ? true : void 0,
    "aria-expanded": context.isOpenSig.value,
    "aria-haspopup": "true",
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HDropdownTrigger = HDropdownTrigger;
