// @qwik-ui/headless/components/dropdown/dropdown-popover.qwik.mjs
import { jsx } from "@builder.io/qwik/jsx-runtime";
import { component$, useStyles$, useContext, useSignal, useTask$, $, Slot } from "@builder.io/qwik";
import { usePopover } from "../popover/use-popover.qwik.mjs";
import { HPopoverPanel } from "../popover/popover-panel.qwik.mjs";
import { dropdownContextId } from "./dropdown-context.qwik.mjs";
import { HPopoverRoot } from "../popover/popover-root.qwik.mjs";
import { isServer } from "@builder.io/qwik/build";
import { useDebouncer } from "../../hooks/use-debouncer.qwik.mjs";
import styles from "./dropdown.css.qwik.mjs";
const HDropdownPopover = component$((props) => {
  useStyles$(styles);
  const context = useContext(dropdownContextId);
  const { showPopover, hidePopover } = usePopover(context.localId);
  const initialLoadSig = useSignal(true);
  const { floating, flip, hover, gutter, ...rest } = props;
  useTask$(async ({ track }) => {
    track(() => context.isOpenSig.value);
    if (isServer) return;
    if (!initialLoadSig.value) {
      if (context.isOpenSig.value) {
        await showPopover();
      } else {
        await hidePopover();
      }
    }
  });
  const menuId = `${context.localId}-panel`;
  const isOutside = $((rect, x, y) => {
    return x < rect.left || x > rect.right || y < rect.top || y > rect.bottom;
  });
  const handleDismiss$ = $(async (e) => {
    if (!context.isOpenSig.value) {
      return;
    }
    if (!context.panelRef.value || !context.triggerRef.value) {
      return;
    }
    const menuRect = context.panelRef.value.getBoundingClientRect();
    const triggerRect = context.triggerRef.value.getBoundingClientRect();
    const { clientX, clientY } = e;
    const isOutsideMenu = await isOutside(menuRect, clientX, clientY);
    const isOutsideTrigger = await isOutside(triggerRect, clientX, clientY);
    if (isOutsideMenu && isOutsideTrigger) {
      context.isOpenSig.value = false;
    }
  });
  useTask$(({ track, cleanup }) => {
    track(() => context.isOpenSig.value);
    if (isServer) return;
    if (context.isOpenSig.value) {
      window.addEventListener("pointerdown", handleDismiss$);
    }
    cleanup(() => {
      window.removeEventListener("pointerdown", handleDismiss$);
    });
  });
  useTask$(() => {
    initialLoadSig.value = false;
  });
  const resetScrollMove = useDebouncer($(() => {
    context.isMouseOverPopupSig.value = false;
  }), 650);
  return /* @__PURE__ */ jsx(HPopoverRoot, {
    floating,
    flip,
    hover,
    gutter,
    "bind:anchor": props["bind:anchor"] ?? context.triggerRef,
    "bind:panel": context.panelRef,
    manual: true,
    id: context.localId,
    style: {
      display: "contents"
    },
    children: /* @__PURE__ */ jsx(HPopoverPanel, {
      id: menuId,
      "data-open": context.isOpenSig.value ? "" : void 0,
      "data-closed": !context.isOpenSig.value ? "" : void 0,
      role: "menu",
      "aria-expanded": context.isOpenSig.value ? "true" : "false",
      onMouseMove$: async () => {
        context.isMouseOverPopupSig.value = true;
        await resetScrollMove();
      },
      onMouseOut$: () => context.isMouseOverPopupSig.value = false,
      onKeyDown$: () => context.isMouseOverPopupSig.value = true,
      ...rest,
      children: /* @__PURE__ */ jsx(Slot, {})
    })
  });
});
export {
  HDropdownPopover
};
