// @qwik-ui/headless/components/dropdown/use-dropdown-item.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const qwik = require("@builder.io/qwik");
const build = require("@builder.io/qwik/build");
const dropdownContext = require("./dropdown-context.qwik.cjs");
const useDropdown = require("./use-dropdown.qwik.cjs");
function useDropdownItem(props) {
  const context = qwik.useContext(dropdownContext.dropdownContextId);
  const { getNextEnabledItemIndex$, getPrevEnabledItemIndex$ } = useDropdown.useDropdown();
  const { _index, disabled, closeOnSelect = true } = props;
  const itemRef = qwik.useSignal();
  const localIndexSig = qwik.useSignal(null);
  const itemId = `${context.localId}-${_index}`;
  const isHighlightedSig = qwik.useComputed$(() => {
    if (disabled) return;
    if (context.highlightedIndexSig.value === _index) {
      itemRef.value?.focus();
      return true;
    } else {
      return false;
    }
  });
  qwik.useTask$(async function getIndexTask() {
    if (_index === void 0) throw Error("Qwik UI: Dropdown component item cannot find its proper index.");
    localIndexSig.value = _index;
  });
  const checkVisibility$ = qwik.$(async (entries) => {
    const [entry] = entries;
    if (isHighlightedSig.value && !entry.isIntersecting) {
      const containerRect = context.panelRef.value?.getBoundingClientRect();
      const itemRect = itemRef.value?.getBoundingClientRect();
      if (!containerRect || !itemRect || context.isMouseOverPopupSig.value) return;
      const offset = itemRect.top - containerRect.top - containerRect.height / 2 + itemRect.height / 2;
      context.panelRef.value?.scrollBy({
        top: document.hasFocus() ? offset : void 0,
        ...context.scrollOptions
      });
    }
  });
  qwik.useTask$(async function navigationTask({ track, cleanup }) {
    track(() => context.highlightedIndexSig.value);
    if (build.isServer || !context.panelRef.value) return;
    if (props._index !== context.highlightedIndexSig.value) return;
    context.highlightedItemRef = itemRef;
    const hasScrollbar = context.panelRef.value.scrollHeight > context.panelRef.value.clientHeight;
    if (!hasScrollbar) {
      return;
    }
    const observer = new IntersectionObserver(checkVisibility$, {
      root: context.panelRef.value,
      threshold: 1
    });
    cleanup(() => observer?.disconnect());
    if (itemRef.value) {
      observer.observe(itemRef.value);
    }
  });
  const handleClick$ = qwik.$(async () => {
    if (disabled || localIndexSig.value === null) return;
    props.onItemSelect?.();
    if (closeOnSelect) {
      context.isOpenSig.value = false;
      context.triggerRef.value?.focus();
    }
  });
  const handlePointerOver$ = qwik.$(() => {
    if (disabled) return;
    if (localIndexSig.value !== null) {
      context.highlightedIndexSig.value = localIndexSig.value;
    }
  });
  const handleKeyDown$ = qwik.$(async (e) => {
    switch (e.key) {
      case "ArrowDown":
        if (context.isOpenSig.value) {
          context.highlightedIndexSig.value = await getNextEnabledItemIndex$(context.highlightedIndexSig.value);
        }
        break;
      case "ArrowUp":
        if (context.isOpenSig.value) {
          context.highlightedIndexSig.value = await getPrevEnabledItemIndex$(context.highlightedIndexSig.value);
        }
        break;
      case "Home":
        if (context.isOpenSig.value) {
          context.highlightedIndexSig.value = await getNextEnabledItemIndex$(-1);
        }
        break;
      case "End":
        if (context.isOpenSig.value) {
          const lastEnabledOptionIndex = await getPrevEnabledItemIndex$(context.itemsMapSig.value.size);
          context.highlightedIndexSig.value = lastEnabledOptionIndex;
        }
        break;
      case "Escape":
        context.triggerRef.value?.focus();
        context.isOpenSig.value = false;
        break;
      case "Tab":
        context.isOpenSig.value = false;
        break;
      case "Enter":
      case " ":
        if (context.isOpenSig.value) {
          props.onItemSelect?.();
          if (closeOnSelect) {
            context.isOpenSig.value = false;
            context.triggerRef.value?.focus();
          }
        }
        break;
    }
  });
  return {
    handleKeyDown$,
    handleClick$,
    handlePointerOver$,
    itemId,
    itemRef,
    isHighlightedSig
  };
}
exports.useDropdownItem = useDropdownItem;
