// @qwik-ui/headless/components/carousel/bullet.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const context = require("./context.qwik.cjs");
const useCarousel = require("./use-carousel.qwik.cjs");
const CarouselBullet = qwik.component$(({ _index, ...props }) => {
  const context$1 = qwik.useContext(context.carouselContextId);
  const bulletRef = qwik.useSignal();
  const slideId = `${context$1.localId}-${_index ?? -1}`;
  const isRenderedSig = qwik.useSignal(true);
  const { validIndexesSig } = useCarousel.useCarousel(context$1);
  qwik.useTask$(function getIndexOrder() {
    if (_index !== void 0) {
      context$1.bulletRefsArray.value[_index] = bulletRef;
    } else {
      throw new Error("Qwik UI: Carousel Bullet cannot find its proper index.");
    }
  });
  const handleClick$ = qwik.$(() => {
    if (typeof _index !== "number") return;
    context$1.currentIndexSig.value = _index;
  });
  const handleFocus$ = qwik.$(() => {
    if (typeof _index !== "number") return;
    context$1.currentIndexSig.value = _index;
  });
  const handleKeyDownSync$ = qwik.sync$((e) => {
    if (e.key === "Home" || e.key === "End") {
      e.preventDefault();
    }
  });
  const handleKeyDown$ = qwik.$(async (e) => {
    const usedKeys = [
      "ArrowRight",
      "ArrowLeft",
      "Home",
      "End"
    ];
    if (typeof _index !== "number" || !usedKeys.includes(e.key)) return;
    const validIndexes = validIndexesSig.value;
    const lastScrollableIndex = validIndexes[validIndexes.length - 1];
    if (e.key === "Home") {
      context$1.currentIndexSig.value = validIndexes[0];
      context$1.bulletRefsArray.value[validIndexes[0]].value.focus();
      return;
    }
    if (e.key === "End") {
      context$1.currentIndexSig.value = lastScrollableIndex;
      context$1.bulletRefsArray.value[lastScrollableIndex].value.focus();
      return;
    }
    const direction = e.key === "ArrowRight" ? 1 : -1;
    const currentPosition = validIndexes.indexOf(_index);
    let newPosition = currentPosition + direction;
    if (context$1.isRewindSig.value) {
      newPosition = (newPosition + validIndexes.length) % validIndexes.length;
    } else {
      newPosition = Math.max(0, Math.min(newPosition, validIndexes.length - 1));
    }
    const newIndex = validIndexes[newPosition];
    context$1.currentIndexSig.value = newIndex;
    context$1.bulletRefsArray.value[newIndex].value.focus();
  });
  qwik.useTask$(async function renderAvailableBullets() {
    if (typeof _index !== "number") return;
    const validIndexes = validIndexesSig.value;
    isRenderedSig.value = validIndexes.includes(_index);
  });
  const isActiveBulletSig = qwik.useComputed$(() => {
    if (typeof _index !== "number" || !validIndexesSig.value.includes(_index)) return false;
    const currentIndex = context$1.currentIndexSig.value;
    const nextBulletIndex = validIndexesSig.value[validIndexesSig.value.indexOf(_index) + 1];
    return currentIndex >= _index && (nextBulletIndex === void 0 || currentIndex < nextBulletIndex);
  });
  return /* @__PURE__ */ jsxRuntime.jsx("button", {
    ref: bulletRef,
    role: "tab",
    hidden: !isRenderedSig.value,
    tabIndex: _index === context$1.currentIndexSig.value ? 0 : -1,
    "aria-label": `${_index !== void 0 && `Slide ${_index + 1}`}`,
    "aria-controls": slideId,
    "data-active": isActiveBulletSig.value ? "" : void 0,
    "aria-selected": isActiveBulletSig.value,
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    onFocus$: [
      handleFocus$,
      props.onFocus$
    ],
    onKeyDown$: [
      handleKeyDownSync$,
      handleKeyDown$,
      props.onKeyDown$
    ],
    ...props,
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.CarouselBullet = CarouselBullet;
