// @qwik-ui/headless/components/carousel/index.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const inline = require("./inline.qwik.cjs");
const scroller = require("./scroller.qwik.cjs");
const slide = require("./slide.qwik.cjs");
const previous = require("./previous.qwik.cjs");
const next = require("./next.qwik.cjs");
const pagination = require("./pagination.qwik.cjs");
const bullet = require("./bullet.qwik.cjs");
const title = require("./title.qwik.cjs");
const player = require("./player.qwik.cjs");
const stepper = require("./stepper.qwik.cjs");
const step = require("./step.qwik.cjs");
exports.Root = inline.CarouselRoot;
exports.Scroller = scroller.CarouselScroller;
exports.Slide = slide.CarouselSlide;
exports.Previous = previous.CarouselPrevious;
exports.Next = next.CarouselNext;
exports.Pagination = pagination.CarouselPagination;
exports.Bullet = bullet.CarouselBullet;
exports.Title = title.CarouselTitle;
exports.Player = player.CarouselPlayer;
exports.Stepper = stepper.CarouselStepper;
exports.Step = step.CarouselStep;
