// @qwik-ui/headless/components/carousel/previous.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const context = require("./context.qwik.cjs");
const useCarousel = require("./use-carousel.qwik.cjs");
const CarouselPrevious = qwik.component$((props) => {
  const context$1 = qwik.useContext(context.carouselContextId);
  const isKeyboardFocusSig = qwik.useSignal(false);
  const { validIndexesSig } = useCarousel.useCarousel(context$1);
  const isFirstScrollableIndexSig = qwik.useComputed$(() => {
    return validIndexesSig.value[0];
  });
  const handleFocusNext$ = qwik.$(() => {
    if (context$1.isRewindSig.value) return;
    if (isKeyboardFocusSig.value && context$1.currentIndexSig.value === isFirstScrollableIndexSig.value) {
      const activeElAtBlur = document.activeElement;
      setTimeout(() => {
        if (document.activeElement !== activeElAtBlur) return;
        if (context$1.currentIndexSig.value === isFirstScrollableIndexSig.value) {
          context$1.nextButtonRef.value?.focus();
        }
      }, 2e3);
    }
    isKeyboardFocusSig.value = false;
  });
  const handleKeyDown$ = qwik.$(() => {
    isKeyboardFocusSig.value = true;
  });
  const handleClick$ = qwik.$(() => {
    const validIndexes = validIndexesSig.value;
    const currentIndex = context$1.currentIndexSig.value;
    const currentPosition = validIndexes.indexOf(currentIndex);
    if (currentPosition === 0 && context$1.isRewindSig.value) {
      context$1.currentIndexSig.value = validIndexes[validIndexes.length - 1];
    } else {
      const prevPosition = Math.max(currentPosition - 1, 0);
      context$1.currentIndexSig.value = validIndexes[prevPosition];
    }
  });
  const isFirstSlideSig = qwik.useComputed$(() => {
    return context$1.currentIndexSig.value === isFirstScrollableIndexSig.value;
  });
  return /* @__PURE__ */ jsxRuntime.jsx("button", {
    ...props,
    ref: context$1.prevButtonRef,
    "aria-disabled": isFirstSlideSig.value && !context$1.isRewindSig.value,
    disabled: isFirstSlideSig.value && !context$1.isRewindSig.value,
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    onBlur$: [
      handleFocusNext$,
      props.onBlur$
    ],
    onKeyDown$: [
      handleKeyDown$,
      props.onKeyDown$
    ],
    "data-qui-carousel-prev": true,
    "aria-label": "previous slide",
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.CarouselPrevious = CarouselPrevious;
