// @qwik-ui/headless/components/carousel/root.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const context = require("./context.qwik.cjs");
const boundSignal = require("../../utils/bound-signal.qwik.cjs");
const useCarousel = require("./use-carousel.qwik.cjs");
const CarouselBase = qwik.component$((props) => {
  const { align, "bind:currSlideIndex": givenOldSlideIndexSig, "bind:selectedIndex": givenSlideIndexSig, "bind:autoplay": givenAutoplaySig, "bind:progress": givenProgressSig, _isTitle: isTitle, startIndex, ...rest } = props;
  const localId = qwik.useId();
  const scrollerRef = qwik.useSignal();
  const nextButtonRef = qwik.useSignal();
  const prevButtonRef = qwik.useSignal();
  const scrollStartRef = qwik.useSignal();
  const isMouseDraggingSig = qwik.useSignal(false);
  const slideRefsArray = qwik.useSignal([]);
  const bulletRefsArray = qwik.useSignal([]);
  const startIndexSig = qwik.useComputed$(() => {
    return startIndex ?? givenSlideIndexSig?.value ?? 0;
  });
  const currentIndexSig = boundSignal.useBoundSignal(givenSlideIndexSig ?? givenOldSlideIndexSig, startIndexSig.value);
  const isScrollerSig = qwik.useSignal(false);
  const isAutoplaySig = boundSignal.useBoundSignal(givenAutoplaySig, false);
  const getInitialProgress = () => {
    return startIndexSig.value ? startIndexSig.value / ((props._numSlides ?? 1) - 1) : 0;
  };
  const numSlidesSig = qwik.useComputed$(() => props._numSlides ?? 0);
  const isDraggableSig = qwik.useComputed$(() => props.draggable ?? true);
  const slidesPerViewSig = qwik.useComputed$(() => props.slidesPerView ?? 1);
  const gapSig = qwik.useComputed$(() => props.gap ?? 0);
  const alignSig = qwik.useComputed$(() => props.align ?? "start");
  const isRewindSig = qwik.useComputed$(() => props.rewind ?? false);
  const autoPlayIntervalMsSig = qwik.useComputed$(() => props.autoPlayIntervalMs ?? 0);
  const progressSig = boundSignal.useBoundSignal(givenProgressSig, getInitialProgress());
  const sensitivitySig = qwik.useComputed$(() => {
    return {
      mouse: props.sensitivity?.mouse ?? 1.5,
      touch: props.sensitivity?.touch ?? 1.25
    };
  });
  const moveSig = qwik.useComputed$(() => props.move ?? 1);
  const maxSlideHeightSig = qwik.useComputed$(() => props.maxSlideHeight ?? void 0);
  const orientationSig = qwik.useComputed$(() => {
    if (props.maxSlideHeight === void 0) {
      return "horizontal";
    }
    return props.orientation ?? "horizontal";
  });
  const isMouseWheelSig = qwik.useComputed$(() => props.mousewheel ?? false);
  const titleId = `${localId}-title`;
  const context$1 = {
    localId,
    scrollerRef,
    nextButtonRef,
    prevButtonRef,
    scrollStartRef,
    isMouseDraggingSig,
    isMouseWheelSig,
    slideRefsArray,
    bulletRefsArray,
    currentIndexSig,
    isScrollerSig,
    isAutoplaySig,
    numSlidesSig,
    isDraggableSig,
    slidesPerViewSig,
    gapSig,
    alignSig,
    isRewindSig,
    autoPlayIntervalMsSig,
    startIndexSig,
    sensitivitySig,
    moveSig,
    orientationSig,
    maxSlideHeightSig
  };
  useCarousel.useAutoplay(context$1);
  qwik.useContextProvider(context.carouselContextId, context$1);
  qwik.useTask$(({ track }) => {
    if (!givenProgressSig) return;
    track(() => currentIndexSig.value);
    track(() => numSlidesSig.value);
    if (numSlidesSig.value > 1) {
      progressSig.value = currentIndexSig.value / (numSlidesSig.value - 1) * 100;
    } else {
      progressSig.value = 0;
    }
  });
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    role: "group",
    "aria-labelledby": isTitle ? titleId : void 0,
    "aria-label": !isTitle ? `content slideshow` : void 0,
    "aria-roledescription": "carousel",
    "aria-live": isAutoplaySig.value ? "off" : "polite",
    "data-qui-carousel": true,
    ...rest,
    style: {
      "--slides-per-view": slidesPerViewSig.value,
      "--gap": `${gapSig.value}px`,
      "--scroll-snap-align": alignSig.value,
      "--orientation": orientationSig.value === "vertical" ? "column" : "row",
      "--max-slide-height": `${maxSlideHeightSig.value}px`
    },
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.CarouselBase = CarouselBase;
