// @qwik-ui/headless/components/carousel/slide.qwik.mjs
import { jsxs, Fragment, jsx } from "@builder.io/qwik/jsx-runtime";
import { component$, useContext, useSignal, useComputed$, useTask$, $, Slot } from "@builder.io/qwik";
import { carouselContextId } from "./context.qwik.mjs";
const CarouselSlide = component$(({ _index, ...props }) => {
  const context = useContext(carouselContextId);
  const slideRef = useSignal();
  const slideId = `${context.localId}-${_index ?? -1}`;
  const isVisibleSig = useComputed$(() => {
    const start = context.currentIndexSig.value;
    const end = start + context.slidesPerViewSig.value;
    return _index !== void 0 && _index >= start && _index < end;
  });
  const isActiveSig = useComputed$(() => {
    return context.currentIndexSig.value === _index;
  });
  const isInactiveSig = useComputed$(() => {
    return !context.isScrollerSig.value && !isActiveSig.value;
  });
  useTask$(function getIndexOrder() {
    if (_index !== void 0) {
      context.slideRefsArray.value[_index] = slideRef;
    } else {
      throw new Error("Qwik UI: Carousel Slide cannot find its proper index.");
    }
  });
  const handleFocusIn$ = $(() => {
    context.isAutoplaySig.value = false;
  });
  return /* @__PURE__ */ jsxs(Fragment, {
    children: [
      /* @__PURE__ */ jsx(ScrollMarker, {
        align: "start",
        _index
      }),
      /* @__PURE__ */ jsxs("div", {
        ref: slideRef,
        id: slideId,
        inert: !isVisibleSig.value,
        hidden: isInactiveSig.value,
        "aria-roledescription": "slide",
        "data-orientation": context.orientationSig.value,
        role: context.bulletRefsArray.value.length > 0 ? "tabpanel" : void 0,
        "data-qui-carousel-slide": true,
        "data-active": isVisibleSig.value ? "" : void 0,
        "aria-label": `${_index !== void 0 && _index + 1} of ${context.numSlidesSig.value}`,
        onFocusIn$: [
          handleFocusIn$,
          props.onFocusIn$
        ],
        ...props,
        children: [
          /* @__PURE__ */ jsx(Slot, {}),
          /* @__PURE__ */ jsx(ScrollMarker, {
            align: "center",
            _index
          })
        ]
      }),
      /* @__PURE__ */ jsx(ScrollMarker, {
        align: "end",
        _index
      })
    ]
  });
});
const ScrollMarker = component$((props) => {
  const context = useContext(carouselContextId);
  const { align, _index, ...rest } = props;
  if (_index === void 0) return;
  const isVisible = context.startIndexSig.value === _index && context.alignSig.value === align;
  return /* @__PURE__ */ jsx("div", {
    ref: context.scrollStartRef,
    "data-qui-scroll-start": true,
    "data-orientation": context.orientationSig.value,
    hidden: !isVisible,
    [`data-${align}`]: "",
    ...rest
  });
});
export {
  CarouselSlide
};
