// @qwik-ui/headless/components/carousel/slide.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const context = require("./context.qwik.cjs");
const CarouselSlide = qwik.component$(({ _index, ...props }) => {
  const context$1 = qwik.useContext(context.carouselContextId);
  const slideRef = qwik.useSignal();
  const slideId = `${context$1.localId}-${_index ?? -1}`;
  const isVisibleSig = qwik.useComputed$(() => {
    const start = context$1.currentIndexSig.value;
    const end = start + context$1.slidesPerViewSig.value;
    return _index !== void 0 && _index >= start && _index < end;
  });
  const isActiveSig = qwik.useComputed$(() => {
    return context$1.currentIndexSig.value === _index;
  });
  const isInactiveSig = qwik.useComputed$(() => {
    return !context$1.isScrollerSig.value && !isActiveSig.value;
  });
  qwik.useTask$(function getIndexOrder() {
    if (_index !== void 0) {
      context$1.slideRefsArray.value[_index] = slideRef;
    } else {
      throw new Error("Qwik UI: Carousel Slide cannot find its proper index.");
    }
  });
  const handleFocusIn$ = qwik.$(() => {
    context$1.isAutoplaySig.value = false;
  });
  return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, {
    children: [
      /* @__PURE__ */ jsxRuntime.jsx(ScrollMarker, {
        align: "start",
        _index
      }),
      /* @__PURE__ */ jsxRuntime.jsxs("div", {
        ref: slideRef,
        id: slideId,
        inert: !isVisibleSig.value,
        hidden: isInactiveSig.value,
        "aria-roledescription": "slide",
        "data-orientation": context$1.orientationSig.value,
        role: context$1.bulletRefsArray.value.length > 0 ? "tabpanel" : void 0,
        "data-qui-carousel-slide": true,
        "data-active": isVisibleSig.value ? "" : void 0,
        "aria-label": `${_index !== void 0 && _index + 1} of ${context$1.numSlidesSig.value}`,
        onFocusIn$: [
          handleFocusIn$,
          props.onFocusIn$
        ],
        ...props,
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {}),
          /* @__PURE__ */ jsxRuntime.jsx(ScrollMarker, {
            align: "center",
            _index
          })
        ]
      }),
      /* @__PURE__ */ jsxRuntime.jsx(ScrollMarker, {
        align: "end",
        _index
      })
    ]
  });
});
const ScrollMarker = qwik.component$((props) => {
  const context$1 = qwik.useContext(context.carouselContextId);
  const { align, _index, ...rest } = props;
  if (_index === void 0) return;
  const isVisible = context$1.startIndexSig.value === _index && context$1.alignSig.value === align;
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    ref: context$1.scrollStartRef,
    "data-qui-scroll-start": true,
    "data-orientation": context$1.orientationSig.value,
    hidden: !isVisible,
    [`data-${align}`]: "",
    ...rest
  });
});
exports.CarouselSlide = CarouselSlide;
