// @qwik-ui/headless/components/carousel/use-scroller.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const qwik = require("@builder.io/qwik");
function useScroller(context) {
  const startPosSig = qwik.useSignal();
  const transformSig = qwik.useSignal({
    x: 0,
    y: 0,
    z: 0
  });
  const boundariesSig = qwik.useSignal(null);
  const isMouseDownSig = qwik.useSignal(false);
  const isTouchDeviceSig = qwik.useSignal(false);
  const isInitialTransitionSig = qwik.useSignal(true);
  const givenTransitionSig = qwik.useSignal();
  qwik.useTask$(() => {
    context.isScrollerSig.value = true;
  });
  const orientationProps = {
    vertical: {
      size: "height",
      scroll: "scrollHeight",
      client: "clientHeight",
      direction: "y",
      pagePosition: "pageY",
      clientPosition: "clientY"
    },
    horizontal: {
      size: "width",
      scroll: "scrollWidth",
      client: "clientWidth",
      direction: "x",
      pagePosition: "pageX",
      clientPosition: "clientX"
    }
  };
  const { direction, scroll, client, size } = orientationProps[context.orientationSig.value];
  const setTransform = qwik.$(() => {
    if (!context.scrollerRef.value) return;
    const transformMap = {
      x: (value2) => `${value2}px, 0, 0`,
      y: (value2) => `0, ${value2}px, 0`
    };
    const value = transformSig.value[direction];
    context.scrollerRef.value.style.transform = `translate3d(${transformMap[direction](value)})`;
  });
  const setBoundaries = qwik.$(() => {
    if (!context.scrollerRef.value) return;
    const maxTransform = 0;
    const minTransform = -(context.scrollerRef.value[scroll] - context.scrollerRef.value[client]);
    boundariesSig.value = {
      min: minTransform,
      max: maxTransform
    };
  });
  const setTransition = qwik.$((isEnabled) => {
    if (!context.scrollerRef.value) return;
    if (isInitialTransitionSig.value) {
      givenTransitionSig.value = getComputedStyle(context.scrollerRef.value).transition;
      isInitialTransitionSig.value = false;
    }
    context.scrollerRef.value.style.transition = isEnabled ? givenTransitionSig.value ?? "revert" : "none";
  });
  const getSlidePosition = qwik.$((index) => {
    if (!context.scrollerRef.value) return 0;
    const container = context.scrollerRef.value;
    const slides = context.slideRefsArray.value;
    let position = 0;
    for (let i = 0; i < index; i++) {
      if (slides[i].value) {
        const rect = slides[i].value.getBoundingClientRect();
        position += rect[size];
        position += context.gapSig.value;
      }
    }
    const alignment = context.alignSig.value;
    const currentSlide = slides[index].value;
    const currentRect = currentSlide.getBoundingClientRect();
    const containerSize = container[client];
    const slideSize = currentRect[size];
    if (alignment === "center") {
      position -= (containerSize - slideSize) / 2;
    } else if (alignment === "end") {
      position -= containerSize - slideSize;
    }
    const maxPosition = 0;
    const minPosition = -(container[scroll] - containerSize);
    position = Math.max(minPosition, Math.min(maxPosition, -position));
    return Math.abs(position);
  });
  const setInitialSlidePos = qwik.$(async () => {
    if (context.startIndexSig.value === void 0) {
      throw new Error("Qwik UI: Q Visible executed when startIndexSig is not set");
    }
    if (!context.scrollerRef.value) {
      return;
    }
    context.scrollerRef.value.style.transform = "none";
    const scrollLeft = context.scrollerRef.value.scrollLeft;
    const scrollTop = context.scrollerRef.value.scrollTop;
    transformSig.value = {
      x: -scrollLeft,
      y: -scrollTop,
      z: transformSig.value.z
    };
    await setTransition(false);
    await setTransform();
    context.scrollerRef.value.style.overflow = "visible";
  });
  return {
    startPosSig,
    transformSig,
    boundariesSig,
    isMouseDownSig,
    isTouchDeviceSig,
    isInitialTransitionSig,
    setTransform,
    setBoundaries,
    setTransition,
    setInitialSlidePos,
    orientationProps,
    getSlidePosition
  };
}
exports.useScroller = useScroller;
