// @qwik-ui/headless/components/combobox/combobox-trigger.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const comboboxContext = require("./combobox-context.qwik.cjs");
const combinedRefs = require("../../hooks/combined-refs.qwik.cjs");
const HComboboxTrigger = qwik.component$((props) => {
  const context = qwik.useContext(comboboxContext.comboboxContextId);
  const contextRefOpts = {
    context,
    givenContextRef: context.triggerRef
  };
  const triggerRef = combinedRefs.useCombinedRef(props.ref, contextRefOpts);
  const handleClick$ = qwik.$(() => {
    context.inputRef.value?.focus();
    context.isListboxOpenSig.value = !context.isListboxOpenSig.value;
  });
  return /* @__PURE__ */ jsxRuntime.jsx("button", {
    ref: triggerRef,
    type: "button",
    "aria-expanded": context.isExpandedSig.value,
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    "data-open": context.isListboxOpenSig.value ? "" : void 0,
    "data-closed": !context.isListboxOpenSig.value ? "" : void 0,
    "data-invalid": context.isInvalidSig.value ? "" : void 0,
    "preventdefault:mousedown": true,
    tabIndex: -1,
    ...props,
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HComboboxTrigger = HComboboxTrigger;
