// @qwik-ui/headless/components/combobox/combobox-hidden-select.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const comboboxContext = require("./combobox-context.qwik.cjs");
const visuallyHidden = require("../../utils/visually-hidden.qwik.cjs");
const comboboxHiddenOption = require("./combobox-hidden-option.qwik.cjs");
const HiddenNativeSelect = qwik.component$((props) => {
  const { label, autoComplete, ref, ...rest } = props;
  const context = qwik.useContext(comboboxContext.comboboxContextId);
  const nativeSelectRef = qwik.useSignal();
  return /* @__PURE__ */ jsxRuntime.jsx(visuallyHidden.VisuallyHidden, {
    children: /* @__PURE__ */ jsxRuntime.jsx("div", {
      "aria-hidden": "true",
      children: /* @__PURE__ */ jsxRuntime.jsxs("label", {
        children: [
          label,
          /* @__PURE__ */ jsxRuntime.jsxs("select", {
            ref: (element) => {
              nativeSelectRef.value = element;
              ref?.(element);
            },
            onFocus$: () => {
              context.inputRef.value?.focus();
              return;
            },
            multiple: context.multiple,
            tabIndex: -1,
            autocomplete: autoComplete,
            disabled: context.isDisabledSig.value ? true : void 0,
            required: context.required,
            name: context.name,
            // height is determined by its children
            style: {
              height: "1px"
            },
            ...rest,
            children: [
              /* @__PURE__ */ jsxRuntime.jsx("option", {}),
              Array.from(context.itemsMapSig.value.entries()).map(([index, item]) => /* @__PURE__ */ jsxRuntime.jsx(comboboxHiddenOption.ComboboxHiddenSelectOption, {
                value: item.value,
                displayValue: item.displayValue,
                nativeSelectRef,
                index
              }, item.value))
            ]
          })
        ]
      })
    })
  });
});
exports.HiddenNativeSelect = HiddenNativeSelect;
