// @qwik-ui/headless/components/combobox/combobox-trigger.qwik.mjs
import { jsx } from "@builder.io/qwik/jsx-runtime";
import { component$, useContext, $, Slot } from "@builder.io/qwik";
import { comboboxContextId } from "./combobox-context.qwik.mjs";
import { useCombinedRef } from "../../hooks/combined-refs.qwik.mjs";
const HComboboxTrigger = component$((props) => {
  const context = useContext(comboboxContextId);
  const contextRefOpts = {
    context,
    givenContextRef: context.triggerRef
  };
  const triggerRef = useCombinedRef(props.ref, contextRefOpts);
  const handleClick$ = $(() => {
    context.inputRef.value?.focus();
    context.isListboxOpenSig.value = !context.isListboxOpenSig.value;
  });
  return /* @__PURE__ */ jsx("button", {
    ref: triggerRef,
    type: "button",
    "aria-expanded": context.isExpandedSig.value,
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    "data-open": context.isListboxOpenSig.value ? "" : void 0,
    "data-closed": !context.isListboxOpenSig.value ? "" : void 0,
    "data-invalid": context.isInvalidSig.value ? "" : void 0,
    "preventdefault:mousedown": true,
    tabIndex: -1,
    ...props,
    children: /* @__PURE__ */ jsx(Slot, {})
  });
});
export {
  HComboboxTrigger
};
