// @qwik-ui/headless/components/combobox/combobox-label.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const comboboxContext = require("./combobox-context.qwik.cjs");
const label = require("../label/label.qwik.cjs");
const combinedRefs = require("../../hooks/combined-refs.qwik.cjs");
const HComboboxLabel = qwik.component$((props) => {
  const context = qwik.useContext(comboboxContext.comboboxContextId);
  const labelId = `${context.localId}-label`;
  const contextRefOpts = {
    context,
    givenContextRef: context.labelRef
  };
  const labelRef = combinedRefs.useCombinedRef(props.ref, contextRefOpts);
  const handleClick$ = qwik.$(() => {
    context.inputRef.value?.focus();
  });
  return /* @__PURE__ */ jsxRuntime.jsx(label.HLabel, {
    onClick$: handleClick$,
    id: labelId,
    ref: labelRef,
    ...props,
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HComboboxLabel = HComboboxLabel;
