// @qwik-ui/headless/components/tooltip/tooltip-root.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const popoverRoot = require("../popover/popover-root.qwik.cjs");
const tooltipContext = require("./tooltip-context.qwik.cjs");
const HTooltipRoot = qwik.component$((props) => {
  const { placement = "top", id, gutter, delayDuration = 0, flip, onOpenChange$, ...rest } = props;
  const triggerRef = qwik.useSignal();
  const tooltipState = qwik.useSignal("closed");
  const localId = qwik.useId();
  const compId = id ?? localId;
  const context = {
    compId,
    localId,
    triggerRef,
    delayDuration,
    state: tooltipState,
    onOpenChange$: qwik.$((e) => onOpenChange$?.(e))
  };
  qwik.useContextProvider(tooltipContext.TooltipContextId, context);
  return /* @__PURE__ */ jsxRuntime.jsx(popoverRoot.HPopoverRoot, {
    manual: true,
    hover: true,
    "bind:anchor": triggerRef,
    floating: placement,
    id: localId,
    gutter,
    flip,
    children: /* @__PURE__ */ jsxRuntime.jsx("div", {
      id: localId,
      ...rest,
      children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
    })
  });
});
exports.HTooltipRoot = HTooltipRoot;
