// @qwik-ui/headless/components/checkbox/checkbox-root.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const boundSignal = require("../../utils/bound-signal.qwik.cjs");
const checkboxContext = require("./checkbox-context.qwik.cjs");
const CheckboxRoot = qwik.component$((props) => {
  const { "bind:checked": givenCheckedSig, initialValue, as, ...rest } = props;
  const Comp = as ?? "div";
  const checkedSignal = boundSignal.useBoundSignal(givenCheckedSig, initialValue);
  qwik.useContextProvider(checkboxContext.CheckboxContext, checkedSignal);
  const handleKeyDownSync$ = qwik.sync$((e) => {
    if (e.key === " ") {
      e.preventDefault();
    }
  });
  const handleClick$ = qwik.$(() => {
    checkedSignal.value = !checkedSignal.value;
  });
  const handleKeyDown$ = qwik.$((e) => {
    if (e.key === " ") {
      checkedSignal.value = !checkedSignal.value;
    }
  });
  return /* @__PURE__ */ jsxRuntime.jsx(jsxRuntime.Fragment, {
    children: /* @__PURE__ */ jsxRuntime.jsx(Comp, {
      ...rest,
      tabIndex: 0,
      role: "checkbox",
      "aria-checked": checkedSignal.value,
      "aria-labelledby": props["aria-labelledby"],
      // need the onClick$ to work with the handleClick$ below
      // onClick$={onClick$ || handleClick$}
      onClick$: handleClick$,
      onKeyDown$: [
        handleKeyDownSync$,
        handleKeyDown$
      ],
      onKeyPress$: handleClick$,
      "data-qds-checkbox-root": true,
      children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
    })
  });
});
exports.CheckboxRoot = CheckboxRoot;
