// @qwik-ui/headless/components/checkbox/checkbox-root.qwik.mjs
import { jsx, Fragment } from "@builder.io/qwik/jsx-runtime";
import { component$, useContextProvider, sync$, $, Slot } from "@builder.io/qwik";
import { useBoundSignal } from "../../utils/bound-signal.qwik.mjs";
import { CheckboxContext } from "./checkbox-context.qwik.mjs";
const CheckboxRoot = component$((props) => {
  const { "bind:checked": givenCheckedSig, initialValue, as, ...rest } = props;
  const Comp = as ?? "div";
  const checkedSignal = useBoundSignal(givenCheckedSig, initialValue);
  useContextProvider(CheckboxContext, checkedSignal);
  const handleKeyDownSync$ = sync$((e) => {
    if (e.key === " ") {
      e.preventDefault();
    }
  });
  const handleClick$ = $(() => {
    checkedSignal.value = !checkedSignal.value;
  });
  const handleKeyDown$ = $((e) => {
    if (e.key === " ") {
      checkedSignal.value = !checkedSignal.value;
    }
  });
  return /* @__PURE__ */ jsx(Fragment, {
    children: /* @__PURE__ */ jsx(Comp, {
      ...rest,
      tabIndex: 0,
      role: "checkbox",
      "aria-checked": checkedSignal.value,
      "aria-labelledby": props["aria-labelledby"],
      // need the onClick$ to work with the handleClick$ below
      // onClick$={onClick$ || handleClick$}
      onClick$: handleClick$,
      onKeyDown$: [
        handleKeyDownSync$,
        handleKeyDown$
      ],
      onKeyPress$: handleClick$,
      "data-qds-checkbox-root": true,
      children: /* @__PURE__ */ jsx(Slot, {})
    })
  });
});
export {
  CheckboxRoot
};
