// @qwik-ui/headless/components/toggle/toggle.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const boundSignal = require("../../utils/bound-signal.qwik.cjs");
const HToggle = qwik.component$((props) => {
  const { pressed: pressedProp, onPressedChange$, "bind:pressed": givenValueSig, ...buttonProps } = props;
  const pressedSig = boundSignal.useBoundSignal(givenValueSig, pressedProp ? pressedProp : false);
  const handleKeyDownSync$ = qwik.sync$((event) => {
    if (![
      "ArrowRight",
      "ArrowLeft",
      "ArrowDown",
      "ArrowUp"
    ].includes(event.key)) return;
    event.preventDefault();
  });
  qwik.useTask$(async ({ track }) => {
    if (pressedProp === void 0) return;
    track(() => pressedProp);
    pressedSig.value = pressedProp;
  });
  const handleClick$ = qwik.$(async () => {
    if (!props.disabled) {
      pressedSig.value = !pressedSig.value;
      if (onPressedChange$) {
        onPressedChange$(pressedSig.value);
      }
    }
  });
  return /* @__PURE__ */ jsxRuntime.jsx("button", {
    type: "button",
    "aria-pressed": pressedSig.value,
    "aria-disabled": props.disabled ? true : false,
    "data-disabled": props.disabled ? "" : void 0,
    onKeyDown$: [
      handleKeyDownSync$,
      props.onKeyDown$
    ],
    ...buttonProps,
    class: props.class,
    onClick$: [
      props.onClick$,
      handleClick$
    ],
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HToggle = HToggle;
