// @qwik-ui/headless/components/accordion/accordion-item.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const collapsible = require("../collapsible/collapsible.qwik.cjs");
const accordionContext = require("./accordion-context.qwik.cjs");
const build = require("@builder.io/qwik/build");
const HAccordionItem = qwik.component$(({ id, _index, value, open, ...props }) => {
  const context = qwik.useContext(accordionContext.accordionContextId);
  const localId = qwik.useId();
  const itemId = id ?? localId + "-item";
  const localIndexSig = qwik.useComputed$(() => {
    return _index ?? -1;
  });
  const isOpenSig = qwik.useSignal(context.initialIndex === localIndexSig.value);
  const initialLoadSig = qwik.useSignal(true);
  const triggerRef = qwik.useSignal(null);
  qwik.useTask$(function internalState({ track }) {
    track(() => context.selectedIndexSig.value);
    if (context.isMultipleSig.value || build.isServer) return;
    if (context.selectedIndexSig.value !== localIndexSig.value) {
      isOpenSig.value = false;
    } else {
      if (!initialLoadSig.value && value && context.onChange$) {
        context.onChange$(value);
      }
      if (!context.givenValueSig) return;
      context.givenValueSig.value = value ?? null;
    }
  });
  qwik.useTask$(function givenReactiveValue({ track }) {
    if (!context.givenValueSig) return;
    track(() => context.givenValueSig?.value);
    if (context.givenValueSig?.value === value) {
      isOpenSig.value = true;
      context.selectedIndexSig.value = localIndexSig.value;
    } else {
      isOpenSig.value = false;
    }
  });
  qwik.useTask$(({ track }) => {
    track(() => localIndexSig.value);
    context.triggerRefsArray.value[localIndexSig.value] = triggerRef;
    track(() => isOpenSig.value);
    if (isOpenSig.value) {
      context.selectedIndexSig.value = localIndexSig.value;
    }
  });
  qwik.useTask$(() => {
    initialLoadSig.value = false;
  });
  const itemContext = {
    isOpenSig,
    localIndexSig,
    triggerRef
  };
  qwik.useContextProvider(accordionContext.accordionItemContextId, itemContext);
  return /* @__PURE__ */ jsxRuntime.jsx(collapsible.HCollapsible, {
    triggerRef,
    "bind:open": open ? void 0 : isOpenSig,
    open: context.isMultipleSig.value ? open : void 0,
    id: itemId,
    disabled: context.disabled || props.disabled,
    collapsible: context.collapsible,
    accordionItem: true,
    ...props,
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HAccordionItem = HAccordionItem;
