// @qwik-ui/headless/components/tabs/tab.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const tabPanel = require("./tab-panel.qwik.cjs");
const tabsContextId = require("./tabs-context-id.qwik.cjs");
const TAB_ID_PREFIX = "_tab_";
const HTab = qwik.component$(({ selectedClassName, tabId, ...props }) => {
  const contextService = qwik.useContext(tabsContextId.tabsContextId);
  const elementRefSig = qwik.useSignal();
  const fullTabElementId = contextService.tabsPrefix + TAB_ID_PREFIX + tabId;
  const fullPanelElementId = contextService.tabsPrefix + tabPanel.TAB_PANEL_ID_PREFIX + tabId;
  const selectedClassNameSig = qwik.useComputed$(() => {
    return selectedClassName || contextService.selectedClassName;
  });
  const isSelectedSig = qwik.useComputed$(() => {
    return contextService.selectedTabIdSig.value === tabId;
  });
  const selectIfAutomatic$ = qwik.$(() => {
    contextService.selectIfAutomatic$(tabId);
  });
  const classNamesSig = qwik.useComputed$(() => [
    // TODO only given class if selected
    isSelectedSig.value && [
      "selected",
      selectedClassNameSig.value
    ]
  ]);
  return /* @__PURE__ */ jsxRuntime.jsx("button", {
    ...props,
    type: "button",
    role: "tab",
    id: fullTabElementId,
    "aria-controls": fullPanelElementId,
    ref: elementRefSig,
    "aria-disabled": props.disabled,
    "data-state": isSelectedSig.value ? "selected" : "unselected",
    onFocus$: [
      selectIfAutomatic$,
      props.onFocus$
    ],
    onMouseEnter$: [
      selectIfAutomatic$,
      props.onMouseEnter$
    ],
    "aria-selected": isSelectedSig.value,
    tabIndex: isSelectedSig.value ? 0 : -1,
    class: [
      props.class,
      classNamesSig.value
    ],
    onClick$: [
      qwik.$(() => contextService.selectTab$(tabId)),
      props.onClick$
    ],
    onKeyDown$: [
      /* KeyCode cannot be used here. */
      qwik.sync$((e) => {
        const keys = [
          "Home",
          "End",
          "PageDown",
          "PageUp",
          "ArrowDown",
          "ArrowUp",
          "ArrowLeft",
          "ArrowRight"
        ];
        if (keys.includes(e.key)) {
          e.preventDefault();
        }
      }),
      qwik.$((e) => contextService.onTabKeyDown$(e.key, tabId)),
      props.onKeyDown$
    ],
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HTab = HTab;
exports.TAB_ID_PREFIX = TAB_ID_PREFIX;
