// @qwik-ui/headless/components/tabs/tabs.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const keyCode_type = require("../../utils/key-code.type.qwik.cjs");
const getEnabledTab = require("./get-enabled-tab.qwik.cjs");
const tab = require("./tab.qwik.cjs");
const tabPanel = require("./tab-panel.qwik.cjs");
const tabsContextId = require("./tabs-context-id.qwik.cjs");
const tabsList = require("./tabs-list.qwik.cjs");
const HTabs = (props) => {
  const { children, tabListComponent: UserTabList, tabComponent: UserTab, tabPanelComponent: UserTabPanel, ...rest } = props;
  const TabList = UserTabList ? UserTabList : tabsList.HTabList;
  const Tab = UserTab ? UserTab : tab.HTab;
  const TabPanel = UserTabPanel ? UserTabPanel : tabPanel.HTabPanel;
  const childrenToProcess = Array.isArray(children) ? [
    ...children
  ] : [
    children
  ];
  let tabListComponent;
  const tabComponents = [];
  const panelComponents = [];
  const tabInfoList = [];
  let panelIndex = 0;
  let selectedIndex;
  while (childrenToProcess.length) {
    const child = childrenToProcess.shift();
    if (!child) {
      continue;
    }
    if (Array.isArray(child)) {
      childrenToProcess.unshift(...child);
      continue;
    }
    switch (child.type) {
      case TabList: {
        tabListComponent = child;
        const tabListChildren = Array.isArray(child.props.children) ? child.props.children : [
          child.props.children
        ];
        childrenToProcess.unshift(...tabListChildren);
        break;
      }
      case Tab: {
        if (child.props.selected) {
          const currentTabIndex = tabComponents.length;
          selectedIndex = currentTabIndex;
          child.props.selected = void 0;
        }
        tabComponents.push(child);
        break;
      }
      case TabPanel: {
        const { label, selected } = child.props;
        const matchedTabComponent = tabComponents[panelIndex];
        const tabIdFromTabMaybe = (matchedTabComponent?.props).tabId || matchedTabComponent?.key;
        const tabId = tabIdFromTabMaybe || child.key || `${panelIndex}`;
        if (label) {
          tabComponents.push(/* @__PURE__ */ jsxRuntime.jsx(Tab, {
            children: label
          }));
        }
        if (selected) {
          selectedIndex = panelIndex;
          child.props.selected = void 0;
        }
        child.key = tabId;
        child.props._tabId = tabId;
        panelComponents.push(child);
        tabInfoList.push({
          tabId,
          index: panelIndex,
          tabProps: {},
          panelProps: child.props
        });
        panelIndex++;
        break;
      }
      default: {
        console.error(`unhandled component ${child.type} given to Tabs`);
      }
    }
  }
  if (tabComponents.length !== panelIndex) {
    console.error(`mismatched number of tabs and panels: ${tabComponents.length} ${panelIndex}`);
  }
  tabComponents.forEach((tab2, index) => {
    const tabId = tabInfoList[index]?.tabId;
    tab2.key = tabId;
    tab2.props.tabId = tabId;
    if (tabInfoList[index].panelProps.disabled !== void 0 && tab2.props.disabled === void 0) {
      tab2.props.disabled = tabInfoList[index].panelProps.disabled;
    }
    tabInfoList[index].tabProps = tab2.props;
  });
  if (tabListComponent) {
    tabListComponent.children = tabComponents;
    tabListComponent.props.children = tabComponents;
  } else {
    tabListComponent = /* @__PURE__ */ jsxRuntime.jsx(TabList, {
      children: tabComponents
    });
  }
  if (typeof selectedIndex === "number") {
    rest.selectedIndex = selectedIndex;
  }
  return /* @__PURE__ */ jsxRuntime.jsxs(TabsImpl, {
    tabInfoList,
    ...rest,
    children: [
      tabListComponent,
      panelComponents
    ]
  });
};
const TabsImpl = qwik.component$((props) => {
  const {
    // We take these out of the props for the DOM element but we must refer
    // to them as e.g. props.tabs for reactivity
    tabInfoList: _0,
    behavior = "manual",
    selectedTabId: _1,
    selectedIndex: _2,
    vertical,
    selectedClassName,
    onSelectedIndexChange$,
    onSelectedTabIdChange$,
    "bind:selectedIndex": givenIndexSig,
    "bind:selectedTabId": givenTabIdSig,
    ...rest
  } = props;
  const tabsPrefix = qwik.useId();
  const ref = qwik.useSignal();
  const initialSelectedIndexSig = qwik.useSignal();
  const selectedIndexSig = givenIndexSig || initialSelectedIndexSig;
  const initialSelectedTabIdSig = qwik.useSignal();
  const selectedTabIdSig = givenTabIdSig || initialSelectedTabIdSig;
  qwik.useTask$(function syncTabsTask({ track }) {
    const tabInfoList = track(() => props.tabInfoList);
    const tabId = selectedTabIdSig.value;
    syncSelectedStateSignals(tabInfoList, selectedIndexSig, selectedTabIdSig, {
      tabIdToSelect: tabId
    }, true);
  });
  qwik.useTask$(function syncPropSelectedIndexTask({ track }) {
    const updatedIndexFromProps = track(() => props.selectedIndex);
    syncSelectedStateSignals(props.tabInfoList, selectedIndexSig, selectedTabIdSig, {
      indexToSelect: updatedIndexFromProps
    });
  });
  qwik.useTask$(function syncSelectedIndexSigTask({ track }) {
    const updatedIndexSignal = track(() => selectedIndexSig.value);
    syncSelectedStateSignals(props.tabInfoList, selectedIndexSig, selectedTabIdSig, {
      indexToSelect: updatedIndexSignal
    });
    if (typeof selectedIndexSig.value !== "undefined") {
      onSelectedIndexChange$?.(selectedIndexSig.value);
    }
  });
  qwik.useTask$(function syncPropSelectedTabIdTask({ track }) {
    const updatedTabIdFromProps = track(() => props.selectedTabId);
    syncSelectedStateSignals(props.tabInfoList, selectedIndexSig, selectedTabIdSig, {
      tabIdToSelect: updatedTabIdFromProps
    });
  });
  qwik.useTask$(function syncSelectedTabIdSigTask({ track }) {
    let updatedTabId = track(() => selectedTabIdSig.value);
    if (typeof updatedTabId !== "string") {
      const tab2 = getEnabledTab.getEnabledTab(props.tabInfoList, 0);
      if (tab2) {
        updatedTabId = tab2.tabId;
      }
    }
    syncSelectedStateSignals(props.tabInfoList, selectedIndexSig, selectedTabIdSig, {
      tabIdToSelect: updatedTabId
    });
    if (typeof selectedTabIdSig.value !== "undefined") {
      onSelectedTabIdChange$?.(selectedTabIdSig.value);
    }
  });
  qwik.useTask$(function callOnSelectedChangeTask({ track }) {
    if (!onSelectedIndexChange$) return;
    const idx = track(() => selectedIndexSig.value);
    if (typeof idx === "number" && idx >= 0) onSelectedIndexChange$(idx);
  });
  const selectTab$ = qwik.$((tabId) => {
    syncSelectedStateSignals(props.tabInfoList, selectedIndexSig, selectedTabIdSig, {
      tabIdToSelect: tabId
    });
  });
  const selectIfAutomatic$ = qwik.$((tabId) => {
    if (behavior === "automatic") {
      selectTab$(tabId);
    }
  });
  const onTabKeyDown$ = qwik.$((key, currentTabId) => {
    const tabsRootElement = ref.value;
    const currentFocusedTabIndex = props.tabInfoList.findIndex((tabData) => tabData.tabId === currentTabId);
    let tabInfo;
    if (!vertical && key === keyCode_type.KeyCode.ArrowRight || vertical && key === keyCode_type.KeyCode.ArrowDown) {
      tabInfo = getEnabledTab.findNextEnabledTab(props.tabInfoList, currentFocusedTabIndex + 1, {
        wrap: true
      });
    } else if (!vertical && key === keyCode_type.KeyCode.ArrowLeft || vertical && key === keyCode_type.KeyCode.ArrowUp) {
      tabInfo = getEnabledTab.findPrevEnabledTab(props.tabInfoList, currentFocusedTabIndex, {
        wrap: true
      });
    } else if (key === keyCode_type.KeyCode.Home || key === keyCode_type.KeyCode.PageUp) {
      tabInfo = getEnabledTab.findNextEnabledTab(props.tabInfoList, 0);
    } else if (key === keyCode_type.KeyCode.End || key === keyCode_type.KeyCode.PageDown) {
      tabInfo = getEnabledTab.findPrevEnabledTab(props.tabInfoList, props.tabInfoList.length);
    }
    if (tabInfo) {
      focusOnTab(tabInfo.index);
    }
    function focusOnTab(index) {
      const tabListElement = tabsRootElement?.children[0];
      const tabToFocusOn = tabListElement?.children[index];
      tabToFocusOn.focus();
    }
  });
  const contextService = {
    selectTab$,
    tabsPrefix,
    onTabKeyDown$,
    selectIfAutomatic$,
    selectedTabIdSig,
    selectedClassName
  };
  qwik.useContextProvider(tabsContextId.tabsContextId, contextService);
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    "data-qui-tabs-root": true,
    ref,
    ...rest,
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
const syncSelectedStateSignals = (tabsInfoList, selectedIndexSig, selectedTabIdSig, { indexToSelect, tabIdToSelect }, ignoreIndexNotFound) => {
  if (tabIdToSelect) {
    indexToSelect = tabsInfoList.findIndex((tabInfo) => tabInfo.tabId === tabIdToSelect);
  }
  if (typeof indexToSelect !== "number") return;
  if (indexToSelect && indexToSelect < 0) {
    if (!ignoreIndexNotFound) {
      return;
    }
    indexToSelect = selectedIndexSig.value;
    if (typeof indexToSelect !== "number") return;
  }
  const tab2 = getEnabledTab.getEnabledTab(tabsInfoList, indexToSelect);
  if (tab2 && (tab2.index !== selectedIndexSig.value || tab2.tabId !== selectedTabIdSig.value)) {
    selectedIndexSig.value = tab2.index;
    selectedTabIdSig.value = tab2.tabId;
  }
};
exports.HTabs = HTabs;
exports.TabsImpl = TabsImpl;
exports.syncSelectedStateSignals = syncSelectedStateSignals;
