// @qwik-ui/headless/components/popover/popover-trigger.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const popoverContext = require("./popover-context.qwik.cjs");
const usePopover = require("./use-popover.qwik.cjs");
const HPopoverTrigger = qwik.component$((props) => {
  const context = qwik.useContext(popoverContext.popoverContextId);
  const triggerId = `${context.compId}-trigger`;
  const panelId = `${context.compId}-panel`;
  const { initPopover$, showPopover, hidePopover, hasPolyfillLoadedSig, isSupportedSig } = usePopover.usePopover(context.compId);
  const handleClick$ = qwik.$(async () => {
    if (context.hover && !isSupportedSig.value) {
      await showPopover();
      return;
    }
    if (isSupportedSig.value) return;
    await initPopover$();
    while (!hasPolyfillLoadedSig.value) {
      await new Promise((resolve) => setTimeout(resolve, 10));
    }
    if (!isSupportedSig.value) {
      if (!context.isOpenSig.value) {
        context.panelRef?.value?.showPopover();
      } else {
        context.panelRef?.value?.hidePopover();
      }
    }
  });
  const handlePointerOver$ = qwik.$(async () => {
    if (!context.hover) return;
    await showPopover();
  });
  const handlePointerOut$ = qwik.$(async () => {
    if (!context.hover) return;
    await hidePopover();
  });
  return /* @__PURE__ */ jsxRuntime.jsx("button", {
    ...props,
    ref: context.triggerRef,
    id: triggerId,
    popovertarget: panelId,
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    onPointerOver$: [
      handlePointerOver$,
      props.onPointerOver$
    ],
    onPointerOut$: [
      handlePointerOut$,
      props.onPointerOut$
    ],
    popoverTargetAction: context.hover ? "show" : void 0,
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HPopoverTrigger = HPopoverTrigger;
