// @qwik-ui/headless/components/popover/popover-trigger.qwik.mjs
import { jsx } from "@builder.io/qwik/jsx-runtime";
import { component$, useContext, $, Slot } from "@builder.io/qwik";
import { popoverContextId } from "./popover-context.qwik.mjs";
import { usePopover } from "./use-popover.qwik.mjs";
const HPopoverTrigger = component$((props) => {
  const context = useContext(popoverContextId);
  const triggerId = `${context.compId}-trigger`;
  const panelId = `${context.compId}-panel`;
  const { initPopover$, showPopover, hidePopover, hasPolyfillLoadedSig, isSupportedSig } = usePopover(context.compId);
  const handleClick$ = $(async () => {
    if (context.hover && !isSupportedSig.value) {
      await showPopover();
      return;
    }
    if (isSupportedSig.value) return;
    await initPopover$();
    while (!hasPolyfillLoadedSig.value) {
      await new Promise((resolve) => setTimeout(resolve, 10));
    }
    if (!isSupportedSig.value) {
      if (!context.isOpenSig.value) {
        context.panelRef?.value?.showPopover();
      } else {
        context.panelRef?.value?.hidePopover();
      }
    }
  });
  const handlePointerOver$ = $(async () => {
    if (!context.hover) return;
    await showPopover();
  });
  const handlePointerOut$ = $(async () => {
    if (!context.hover) return;
    await hidePopover();
  });
  return /* @__PURE__ */ jsx("button", {
    ...props,
    ref: context.triggerRef,
    id: triggerId,
    popovertarget: panelId,
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    onPointerOver$: [
      handlePointerOver$,
      props.onPointerOver$
    ],
    onPointerOut$: [
      handlePointerOut$,
      props.onPointerOut$
    ],
    popoverTargetAction: context.hover ? "show" : void 0,
    children: /* @__PURE__ */ jsx(Slot, {})
  });
});
export {
  HPopoverTrigger
};
