// @qwik-ui/headless/components/collapsible/collapsible-content.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const collapsibleContext = require("./collapsible-context.qwik.cjs");
const build = require("@builder.io/qwik/build");
const HCollapsibleContent = qwik.component$((props) => {
  const context = qwik.useContext(collapsibleContext.collapsibleContextId);
  const isHiddenSig = qwik.useSignal(!context.isOpenSig.value);
  const isAnimatedSig = qwik.useSignal(false);
  const contentId = `${context.itemId}-content`;
  const triggerId = `${context.itemId}-trigger`;
  const hideContent$ = qwik.$(() => {
    if (!context.isOpenSig.value) {
      isHiddenSig.value = true;
    }
  });
  qwik.useTask$(async function automaticAnimations({ track }) {
    track(() => context.isOpenSig.value);
    if (build.isServer || !context.contentRef.value) {
      return;
    }
    await context.getContentDimensions$();
    if (context.isOpenSig.value) {
      context.contentRef.value.removeAttribute("data-closed");
      context.contentRef.value.dataset.open = "";
      isHiddenSig.value = false;
    } else {
      context.contentRef.value.dataset.closed = "";
      context.contentRef.value.removeAttribute("data-open");
    }
    const { animationDuration, transitionDuration } = getComputedStyle(context.contentRef.value);
    if (animationDuration !== "0s" || transitionDuration !== "0s") {
      isAnimatedSig.value = true;
    } else {
      isAnimatedSig.value = false;
    }
  });
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    ...props,
    ref: context.contentRef,
    id: contentId,
    "data-collapsible-content": true,
    "data-disabled": context.disabled ? "" : void 0,
    onAnimationEnd$: [
      hideContent$,
      props.onAnimationEnd$
    ],
    onTransitionEnd$: [
      hideContent$,
      props.onTransitionEnd$
    ],
    hidden: isAnimatedSig.value ? isHiddenSig.value : !context.isOpenSig.value,
    "aria-labelledby": triggerId,
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HCollapsibleContent = HCollapsibleContent;
