// @qwik-ui/headless/components/collapsible/collapsible-trigger.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const collapsibleContext = require("./collapsible-context.qwik.cjs");
const HCollapsibleTrigger = qwik.component$(({ onClick$, ...props }) => {
  const context = qwik.useContext(collapsibleContext.collapsibleContextId);
  const contentId = `${context.itemId}-content`;
  const triggerId = `${context.itemId}-trigger`;
  const handleClick$ = qwik.$(async () => {
    if (context.isOpenSig.value && context.collapsible === false) return;
    context.isOpenSig.value = !context.isOpenSig.value;
  });
  qwik.useOnWindow("resize", context.getContentDimensions$);
  return /* @__PURE__ */ jsxRuntime.jsx("button", {
    ...props,
    id: triggerId,
    ref: context.triggerRef,
    disabled: context.disabled,
    "data-disabled": context.disabled ? "" : void 0,
    "aria-disabled": context.disabled ? "true" : "false",
    "data-open": context.isOpenSig.value ? "" : void 0,
    "data-closed": !context.isOpenSig.value ? "" : void 0,
    "aria-expanded": context.isOpenSig.value,
    "aria-controls": contentId,
    onClick$: [
      handleClick$,
      onClick$
    ],
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HCollapsibleTrigger = HCollapsibleTrigger;
