// @qwik-ui/headless/components/pagination/pagination.qwik.mjs
import { jsxs, jsx } from "@builder.io/qwik/jsx-runtime";
import { usePagination } from "./use-pagination.qwik.mjs";
import { component$, Slot } from "@builder.io/qwik";
const HPagination = component$((props) => {
  const { selectedPage, totalPages, onPageChange$, siblingCount, hidePrevButton = false, hideNextButton = false, disabled = false, customArrowTexts: { previous: previousButtonLabel = "PREV", next: nextButtonLabel = "NEXT" } = {}, defaultClass = "", selectedClass = "", dividerClass = "", nextButtonClass = "", prevButtonClass = "", ...rest } = props;
  const isPrevButtonEnabled = () => !hidePrevButton && selectedPage > 1;
  const isNextButtonEnabled = () => !hideNextButton && selectedPage !== totalPages;
  const visibleItems = usePagination(totalPages, selectedPage, siblingCount || 1);
  return /* @__PURE__ */ jsxs("nav", {
    "aria-label": "pagination",
    "data-testid": "pagination",
    ...rest,
    children: [
      /* @__PURE__ */ jsxs("button", {
        class: nextButtonClass,
        "aria-label": "prevAriaLabel",
        disabled: disabled || !isPrevButtonEnabled(),
        onClick$: () => {
          if (selectedPage > 1) {
            onPageChange$(selectedPage - 1);
          }
        },
        children: [
          /* @__PURE__ */ jsx(Slot, {
            name: "prefix"
          }),
          /* @__PURE__ */ jsx("span", {
            children: previousButtonLabel
          })
        ]
      }),
      visibleItems.map((item, index) => {
        const isSelected = selectedPage === item;
        return /* @__PURE__ */ jsx("span", {
          children: typeof item === "string" ? /* @__PURE__ */ jsx("button", {
            class: dividerClass,
            disabled: true,
            children: "..."
          }) : /* @__PURE__ */ jsx("button", {
            class: [
              defaultClass,
              selectedPage === item && selectedClass
            ],
            "aria-label": `Page ${item} of ${totalPages}`,
            "aria-current": selectedPage === item,
            disabled: disabled || isSelected,
            onClick$: () => {
              onPageChange$(item);
            },
            children: item
          })
        }, index);
      }),
      /* @__PURE__ */ jsxs("button", {
        class: prevButtonClass,
        "aria-label": "nextAriaLabel",
        disabled: disabled || !isNextButtonEnabled(),
        onClick$: () => {
          if (selectedPage < totalPages) {
            onPageChange$(selectedPage + 1);
          }
        },
        children: [
          /* @__PURE__ */ jsx("span", {
            children: nextButtonLabel
          }),
          /* @__PURE__ */ jsx(Slot, {
            name: "suffix"
          })
        ]
      })
    ]
  });
});
export {
  HPagination
};
