// @qwik-ui/headless/components/toggle-group/toggle-group-root.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const toggleGroupContext = require("./toggle-group-context.qwik.cjs");
const useToggle = require("./use-toggle.qwik.cjs");
const build = require("@builder.io/qwik/build");
const toggleGroup = require("./toggle-group.css.qwik.cjs");
const HToggleGroupRoot = qwik.component$((props) => {
  qwik.useStyles$(toggleGroup);
  const { onChange$: _, disabled = false, orientation = "horizontal", direction = "ltr", loop = false, ...divProps } = props;
  const commonProps = {
    role: "group",
    "aria-orientation": orientation,
    dir: direction
  };
  const api = useToggle.useToggleGroup(props);
  const rootApiContext = {
    rootId: api.rootId,
    rootOrientation: orientation,
    rootDirection: direction,
    rootIsDisabled: disabled,
    rootIsLoopEnabled: loop,
    rootMultiple: api.multiple,
    activateItem$: api.activateItem$,
    deActivateItem$: api.deActivateItem$,
    getAllItem$: api.getAllItems$,
    pressedValuesSig: api.pressedValuesSig,
    getAndSetTabIndexItem$: api.getAndSetTabIndexItem$,
    registerItem$: api.registerItem$,
    itemsCSR: api.itemsCSR
  };
  const setTabIndexInSSR = qwik.$(async () => {
    const allItems = await rootApiContext.getAllItem$();
    const currentPressedItems = allItems.filter((item) => item.isPressed.value === true);
    if (currentPressedItems.length > 0) {
      currentPressedItems.forEach(async (item) => {
        await rootApiContext.getAndSetTabIndexItem$(item.itemId, 0);
      });
      allItems.filter((item) => item.isPressed.value === false).forEach(async (item) => {
        await rootApiContext.getAndSetTabIndexItem$(item.itemId, -1);
      });
      return;
    }
    const firstNotDisabledItem = allItems.find((item) => item.isDisabled === false);
    if (currentPressedItems.length === 0 && firstNotDisabledItem !== void 0) {
      if (firstNotDisabledItem !== void 0) {
        await rootApiContext.getAndSetTabIndexItem$(firstNotDisabledItem.itemId, 0);
      }
      allItems.filter((item) => item.itemId !== firstNotDisabledItem.itemId).forEach(async (item) => {
        await rootApiContext.getAndSetTabIndexItem$(item.itemId, -1);
      });
      return;
    }
  });
  const setTabIndexInCSR = qwik.$(async () => {
    const allItems = await rootApiContext.getAllItem$();
    const currentPressedItems = allItems.filter((item) => item.isPressed.value === true);
    if (currentPressedItems.length > 0) {
      currentPressedItems.forEach(async (item) => {
        const pressedItem = allItems.find((i) => i.itemId === item.itemId);
        if (!pressedItem) throw "Item Not Found";
        if (pressedItem.ref.value) {
          pressedItem.ref.value.tabIndex = 0;
        }
      });
      allItems.filter((item) => item.isPressed.value === false).forEach(async (item) => {
        const notPressedItem = allItems.find((i) => i.itemId === item.itemId);
        if (!notPressedItem) throw "Item Not Found";
        if (notPressedItem.ref.value) {
          notPressedItem.ref.value.tabIndex = -1;
        }
      });
      return;
    }
    rootApiContext.itemsCSR.value = Array.from(document.querySelectorAll(`.toggle-group-item-${rootApiContext.rootId}`));
    const firstNotDisabledItem = rootApiContext.itemsCSR.value.find((item) => item.ariaDisabled === "false");
    if (currentPressedItems.length === 0 && firstNotDisabledItem !== void 0) {
      if (firstNotDisabledItem !== void 0) {
        firstNotDisabledItem.tabIndex = 0;
      }
      allItems.filter((item) => item.itemId !== firstNotDisabledItem.id).forEach(async (item) => {
        const otherItem = allItems.find((i) => i.itemId === item.itemId);
        if (!otherItem) throw "Item Not Found";
        if (otherItem.ref.value) {
          otherItem.ref.value.tabIndex = -1;
        }
      });
      return;
    }
  });
  qwik.useTask$(async ({ track }) => {
    track(() => api.pressedValuesSig.value);
    if (build.isServer) {
      await setTabIndexInSSR();
    }
    if (build.isBrowser) {
      await setTabIndexInCSR();
    }
  });
  qwik.useContextProvider(toggleGroupContext.toggleGroupRootApiContextId, rootApiContext);
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    ...divProps,
    ...commonProps,
    "data-qui-togglegroup-root": true,
    "data-orientation": orientation,
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HToggleGroupRoot = HToggleGroupRoot;
