// @qwik-ui/headless/components/toggle-group/use-toggle.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const qwik = require("@builder.io/qwik");
const boundSignal = require("../../utils/bound-signal.qwik.cjs");
function useRootItemsRepo() {
  const items = qwik.useSignal(/* @__PURE__ */ new Map());
  const rootId = qwik.useId();
  const itemsCSR = qwik.useSignal([]);
  const registerItem$ = qwik.$((itemId, itemSig) => {
    items.value = items.value.set(itemId, itemSig);
  });
  const getAndSetTabIndexItem$ = qwik.$((itemId, tabIndexValue) => {
    const itemSig = items.value.get(itemId);
    if (!itemSig) throw "Item Not Found";
    if (itemSig) {
      itemSig.value.tabIndex.value = tabIndexValue;
    }
  });
  const getAllItems$ = qwik.$(() => Array.from(items.value.values()).map((signal) => signal.value));
  return {
    getAllItems$,
    getAndSetTabIndexItem$,
    registerItem$,
    rootId,
    itemsCSR
  };
}
function useCreateSingleToggleGroup(props) {
  const { multiple = false, value, onChange$, "bind:value": givenValueSig } = props;
  const pressedValuesSig = boundSignal.useBoundSignal(givenValueSig, value);
  const rootItemsRepo = useRootItemsRepo();
  const handleValueChange$ = qwik.$((newValue) => {
    pressedValuesSig.value = newValue;
    if (onChange$) onChange$(pressedValuesSig.value);
  });
  const activateItem$ = qwik.$((itemValue) => handleValueChange$(itemValue));
  const deActivateItem$ = qwik.$(() => handleValueChange$(""));
  return {
    multiple,
    pressedValuesSig,
    activateItem$,
    deActivateItem$,
    getAllItems$: rootItemsRepo.getAllItems$,
    getAndSetTabIndexItem$: rootItemsRepo.getAndSetTabIndexItem$,
    registerItem$: rootItemsRepo.registerItem$,
    rootId: rootItemsRepo.rootId,
    itemsCSR: rootItemsRepo.itemsCSR
  };
}
function useCreateMultipleToggleGroup(props) {
  const { multiple = true, "bind:value": givenValueSig, value, onChange$ } = props;
  const pressedValuesSig = boundSignal.useBoundSignal(givenValueSig, value || []);
  const rootItemsRepo = useRootItemsRepo();
  const handleValueChange$ = qwik.$((newValue) => {
    pressedValuesSig.value = newValue;
    if (onChange$) onChange$(pressedValuesSig.value);
  });
  const activateItem$ = qwik.$((itemValue) => handleValueChange$([
    ...pressedValuesSig.value,
    itemValue
  ]));
  const deActivateItem$ = qwik.$((itemValue) => handleValueChange$(pressedValuesSig.value.filter((value2) => value2 !== itemValue)));
  return {
    multiple,
    pressedValuesSig,
    activateItem$,
    deActivateItem$,
    getAllItems$: rootItemsRepo.getAllItems$,
    getAndSetTabIndexItem$: rootItemsRepo.getAndSetTabIndexItem$,
    registerItem$: rootItemsRepo.registerItem$,
    rootId: rootItemsRepo.rootId,
    itemsCSR: rootItemsRepo.itemsCSR
  };
}
function isSingleProps(props) {
  return props.multiple === void 0 || props.multiple === false;
}
function useToggleGroup(props) {
  if (isSingleProps(props)) {
    return useCreateSingleToggleGroup(props);
  }
  return useCreateMultipleToggleGroup(props);
}
exports.useToggleGroup = useToggleGroup;
