// @qwik-ui/headless/components/select/select-popover.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const usePopover = require("../popover/use-popover.qwik.cjs");
const popoverPanel = require("../popover/popover-panel.qwik.cjs");
const selectContext = require("./select-context.qwik.cjs");
const popoverRoot = require("../popover/popover-root.qwik.cjs");
const build = require("@builder.io/qwik/build");
const combinedRefs = require("../../hooks/combined-refs.qwik.cjs");
const HSelectPopover = qwik.component$((props) => {
  const context = qwik.useContext(selectContext.default);
  const { showPopover, hidePopover } = usePopover.usePopover(context.localId);
  const contextRefOpts = {
    context,
    givenContextRef: context.popoverRef
  };
  const panelRef = combinedRefs.useCombinedRef(props.ref, contextRefOpts);
  const { floating, flip, hover, gutter, ...rest } = props;
  const initialLoadSig = qwik.useSignal(true);
  qwik.useTask$(async ({ track }) => {
    track(() => context.isListboxOpenSig.value);
    if (build.isServer) return;
    if (!initialLoadSig.value) {
      if (context.isListboxOpenSig.value) {
        showPopover();
      } else {
        hidePopover();
      }
    }
  });
  const listboxId = `${context.localId}-panel`;
  const triggerId = `${context.localId}-trigger`;
  const isOutside = qwik.$((rect, x, y) => {
    return x < rect.left || x > rect.right || y < rect.top || y > rect.bottom;
  });
  const handleDismiss$ = qwik.$(async (e) => {
    if (!context.isListboxOpenSig.value) {
      return;
    }
    if (!context.popoverRef.value || !context.triggerRef.value) {
      return;
    }
    const listboxRect = context.popoverRef.value.getBoundingClientRect();
    const boxRect = context.triggerRef.value.getBoundingClientRect();
    const { clientX, clientY } = e;
    const isOutsideListbox = await isOutside(listboxRect, clientX, clientY);
    const isOutsideBox = await isOutside(boxRect, clientX, clientY);
    if (isOutsideListbox && isOutsideBox) {
      context.isListboxOpenSig.value = false;
    }
  });
  qwik.useTask$(({ track, cleanup }) => {
    track(() => context.isListboxOpenSig.value);
    if (build.isServer) return;
    if (context.isListboxOpenSig.value) {
      window.addEventListener("pointerdown", handleDismiss$);
    }
    cleanup(() => {
      window.removeEventListener("pointerdown", handleDismiss$);
    });
  });
  qwik.useTask$(() => {
    initialLoadSig.value = false;
  });
  const handleMouseEnter$ = qwik.$(() => {
    context.isKeyboardFocusSig.value = false;
    context.isMouseOverPopupSig.value = true;
  });
  return /* @__PURE__ */ jsxRuntime.jsx(popoverRoot.HPopoverRoot, {
    floating,
    flip,
    hover,
    gutter,
    "bind:anchor": props["bind:anchor"] ?? context.triggerRef,
    "bind:panel": panelRef,
    manual: true,
    id: context.localId,
    children: /* @__PURE__ */ jsxRuntime.jsx(popoverPanel.HPopoverPanel, {
      id: listboxId,
      "data-open": context.isListboxOpenSig.value ? "" : void 0,
      "data-closed": !context.isListboxOpenSig.value ? "" : void 0,
      "data-invalid": context.isInvalidSig?.value ? "" : void 0,
      role: "listbox",
      "aria-expanded": context.isListboxOpenSig.value ? "true" : void 0,
      "aria-multiselectable": context.multiple ? "true" : void 0,
      "aria-labelledby": triggerId,
      onMouseEnter$: [
        handleMouseEnter$,
        props.onMouseEnter$
      ],
      ...rest,
      children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
    })
  });
});
exports.HSelectPopover = HSelectPopover;
