// @qwik-ui/headless/components/select/select-root.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const selectContext = require("./select-context.qwik.cjs");
const useSelect = require("./use-select.qwik.cjs");
const combinedRefs = require("../../hooks/combined-refs.qwik.cjs");
const HSelectImpl = qwik.component$((props) => {
  const { _itemsMap, _valuePropIndex: givenValuePropIndex, onChange$, onOpenChange$, scrollOptions: givenScrollOptions, loop: givenLoop, multiple = false, _label, name, required, disabled, invalid, ...rest } = props;
  const rootRef = combinedRefs.useCombinedRef(props.ref);
  const triggerRef = qwik.useSignal();
  const popoverRef = qwik.useSignal();
  const listboxRef = qwik.useSignal();
  const labelRef = qwik.useSignal();
  const groupRef = qwik.useSignal();
  const loop = givenLoop ?? false;
  const localId = qwik.useId();
  const itemsMapSig = qwik.useComputed$(() => {
    return _itemsMap;
  });
  const selectedIndexSetSig = qwik.useSignal(new Set([
    givenValuePropIndex ?? []
  ].flat()));
  const highlightedIndexSig = qwik.useSignal(givenValuePropIndex ?? null);
  const isListboxOpenSig = qwik.useSignal(false);
  const scrollOptions = givenScrollOptions ?? {
    behavior: "smooth",
    block: "center",
    inline: "nearest"
  };
  const currDisplayValueSig = qwik.useSignal();
  const initialLoadSig = qwik.useSignal(true);
  const highlightedItemRef = qwik.useSignal();
  const isDisabledSig = qwik.useSignal(disabled ?? false);
  const isInvalidSig = qwik.useSignal(props.invalid ?? false);
  const isKeyboardFocusSig = qwik.useSignal(false);
  const isMouseOverPopupSig = qwik.useSignal(false);
  qwik.useTask$(({ track }) => {
    isInvalidSig.value = track(() => props.invalid ?? false);
  });
  const context = {
    itemsMapSig,
    currDisplayValueSig,
    triggerRef,
    popoverRef,
    listboxRef,
    labelRef,
    groupRef,
    highlightedItemRef,
    localId,
    highlightedIndexSig,
    selectedIndexSetSig,
    isKeyboardFocusSig,
    isMouseOverPopupSig,
    isListboxOpenSig,
    scrollOptions,
    loop,
    multiple,
    name,
    required,
    isDisabledSig,
    isInvalidSig
  };
  qwik.useContextProvider(selectContext.default, context);
  const { selectionManager$ } = useSelect.useSelect();
  qwik.useTask$(async function reactiveUserValue({ track }) {
    const bindValueSig = props["bind:value"];
    if (!bindValueSig) return;
    track(() => bindValueSig.value);
    for (const [index, item] of itemsMapSig.value) {
      if (bindValueSig.value?.includes(item.value)) {
        await selectionManager$(index, "add");
        if (initialLoadSig.value) {
          context.highlightedIndexSig.value = index;
        }
      } else {
        await selectionManager$(index, "remove");
      }
    }
  });
  qwik.useTask$(function reactiveUserOpen({ track }) {
    const bindOpenSig = props["bind:open"];
    if (!bindOpenSig) return;
    track(() => bindOpenSig.value);
    isListboxOpenSig.value = bindOpenSig.value ?? isListboxOpenSig.value;
  });
  qwik.useTask$(function onOpenChangeTask({ track }) {
    track(() => isListboxOpenSig.value);
    if (!initialLoadSig.value) {
      onOpenChange$?.(isListboxOpenSig.value);
    }
  });
  qwik.useTask$(async function updateConsumerProps({ track }) {
    const bindValueSig = props["bind:value"];
    const bindDisplayTextSig = props["bind:displayValue"];
    track(() => selectedIndexSetSig.value);
    const values = [];
    const displayValues = [];
    for (const index of context.selectedIndexSetSig.value) {
      const item = context.itemsMapSig.value.get(index);
      if (item) {
        values.push(item.value);
        displayValues.push(item.displayValue);
      }
    }
    if (onChange$ && selectedIndexSetSig.value.size > 0) {
      await onChange$(context.multiple ? values : values[0]);
    }
    if (bindValueSig && bindValueSig.value) {
      const currUserSigValues = JSON.stringify(bindValueSig.value);
      const newUserSigValues = JSON.stringify(values);
      if (currUserSigValues !== newUserSigValues) {
        if (context.multiple) {
          bindValueSig.value = values;
        } else {
          bindValueSig.value = values[0];
        }
      }
    }
    context.currDisplayValueSig.value = displayValues;
    if (bindDisplayTextSig && context.currDisplayValueSig.value) {
      bindDisplayTextSig.value = context.multiple ? context.currDisplayValueSig.value : context.currDisplayValueSig.value[0];
    }
  });
  qwik.useTask$(() => {
    initialLoadSig.value = false;
  });
  qwik.useTask$(({ track }) => {
    isDisabledSig.value = track(() => disabled ?? false);
  });
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    role: "group",
    ref: rootRef,
    "data-open": context.isListboxOpenSig.value ? "" : void 0,
    "data-closed": !context.isListboxOpenSig.value ? "" : void 0,
    "data-disabled": isDisabledSig.value ? "" : void 0,
    "data-invalid": context.isInvalidSig?.value ? "" : void 0,
    "aria-invalid": context.isInvalidSig?.value,
    "data-qui-select-root": true,
    ...rest,
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HSelectImpl = HSelectImpl;
