// @qwik-ui/headless/components/select/select-label.qwik.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const jsxRuntime = require("@builder.io/qwik/jsx-runtime");
const qwik = require("@builder.io/qwik");
const selectContext = require("./select-context.qwik.cjs");
const combinedRefs = require("../../hooks/combined-refs.qwik.cjs");
const HSelectLabel = qwik.component$((props) => {
  const context = qwik.useContext(selectContext.default);
  const labelId = `${context.localId}-label`;
  const contextRefOpts = {
    context,
    givenContextRef: context.labelRef
  };
  const labelRef = combinedRefs.useCombinedRef(props.ref, contextRefOpts);
  const handleClick$ = qwik.$(() => {
    if (context.isDisabledSig.value) return;
    context.triggerRef.value?.focus();
  });
  const handleMouseDownSync$ = qwik.sync$((e) => {
    if (!e.defaultPrevented && e.detail > 1) e.preventDefault();
  });
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    "data-disabled": context.isDisabledSig.value ? "" : void 0,
    ref: labelRef,
    id: labelId,
    onClick$: [
      handleClick$,
      props.onClick$
    ],
    onMouseDown$: [
      handleMouseDownSync$,
      props.onMouseDown$
    ],
    ...props,
    children: /* @__PURE__ */ jsxRuntime.jsx(qwik.Slot, {})
  });
});
exports.HSelectLabel = HSelectLabel;
